{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "03c04ccd-b8d8-423c-b7d3-ba812d1391a7",
   "metadata": {},
   "source": [
    "---\n",
    "description: We define all the usefull functions to predict the score from a given\n",
    "  sequence\n",
    "output-file: predictions.html\n",
    "title: predictions\n",
    "\n",
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ae468007-6e88-48b1-8b0c-298a459b7a70",
   "metadata": {},
   "outputs": [],
   "source": [
    "#| hide\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9ab16953-30dd-46c3-9ee0-eae1bedbf1cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "#| default_exp predictions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f6f95cf2-5340-4818-8b9e-246fea3f7879",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/regnier/.local/lib/python3.10/site-packages/sklearn/base.py:440: InconsistentVersionWarning: Trying to unpickle estimator LogisticRegression from version 1.7.1 when using version 1.7.0. This might lead to breaking code or invalid results. Use at your own risk. For more info please refer to:\n",
      "https://scikit-learn.org/stable/model_persistence.html#security-maintainability-limitations\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "#| export\n",
    "import pickle\n",
    "import sklearn\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from dgrec.example_data import get_example_data_dir\n",
    "import RNA\n",
    "from concurrent.futures import ProcessPoolExecutor\n",
    "import os\n",
    "import matplotlib.pyplot as plt\n",
    "from dgrec import encoding"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0af10dfc-65cc-4e39-a9a7-56a9b72ad7fa",
   "metadata": {},
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: '/home/regnier/.local/lib/python3.10/site-packages/dgrec/example_data/e_TRSpAvd_classifier.pkl'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[2], line 10\u001b[0m\n\u001b[1;32m      8\u001b[0m model_name2\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124me_TRSpAvd_classifier.pkl\u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[1;32m      9\u001b[0m model_path2\u001b[38;5;241m=\u001b[39mos\u001b[38;5;241m.\u001b[39mpath\u001b[38;5;241m.\u001b[39mjoin(data_path,model_name2)\n\u001b[0;32m---> 10\u001b[0m model_Avd_Sp\u001b[38;5;241m=\u001b[39mpickle\u001b[38;5;241m.\u001b[39mload(\u001b[38;5;28;43mopen\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43mmodel_path2\u001b[49m\u001b[43m,\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mrb\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m)\n\u001b[1;32m     12\u001b[0m model_name_whole\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mwhole_model.pkl\u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[1;32m     13\u001b[0m model_path_whole\u001b[38;5;241m=\u001b[39mos\u001b[38;5;241m.\u001b[39mpath\u001b[38;5;241m.\u001b[39mjoin(data_path,model_name_whole)\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/IPython/core/interactiveshell.py:324\u001b[0m, in \u001b[0;36m_modified_open\u001b[0;34m(file, *args, **kwargs)\u001b[0m\n\u001b[1;32m    317\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m file \u001b[38;5;129;01min\u001b[39;00m {\u001b[38;5;241m0\u001b[39m, \u001b[38;5;241m1\u001b[39m, \u001b[38;5;241m2\u001b[39m}:\n\u001b[1;32m    318\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\n\u001b[1;32m    319\u001b[0m         \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mIPython won\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mt let you open fd=\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mfile\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m by default \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    320\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mas it is likely to crash IPython. If you know what you are doing, \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    321\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124myou can use builtins\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m open.\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    322\u001b[0m     )\n\u001b[0;32m--> 324\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mio_open\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfile\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: '/home/regnier/.local/lib/python3.10/site-packages/dgrec/example_data/e_TRSpAvd_classifier.pkl'"
     ]
    }
   ],
   "source": [
    "# | export\n",
    "\n",
    "data_path=get_example_data_dir()\n",
    "model_name='e_TRSp_classifier.pkl'\n",
    "model_path=os.path.join(data_path,model_name)\n",
    "model_Sp=pickle.load(open(model_path,\"rb\"))\n",
    "\n",
    "model_name2='e_TRSpAvd_classifier.pkl'\n",
    "model_path2=os.path.join(data_path,model_name2)\n",
    "model_Avd_Sp=pickle.load(open(model_path2,\"rb\"))\n",
    "\n",
    "model_name_whole='whole_model.pkl'\n",
    "model_path_whole=os.path.join(data_path,model_name_whole)\n",
    "model_whole=pickle.load(open(model_path_whole,\"rb\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16189738-bc23-4d9b-af5f-1694070dcbeb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# | export\n",
    "\n",
    "def score(TR_seq:str #A string of the TR DNA sequence\n",
    ",features=1 #The classifier model, no need to specify it (one feature by default). If two: uses the two features model\n",
    "         ):\n",
    "\n",
    "    \"\"\"Calculates the predicted score of a given TR sequence (1 = perfect TR and 0 = crappy TR).  If features=2, returns the score according to each feature (better to have both high).\"\"\"\n",
    "    encoded_TR=encoding.encode_tr_list([TR_seq],features)\n",
    "    if features == 1:\n",
    "        score=np.round(model_Sp.predict_proba([encoded_TR[0]])[:,1],decimals=2).item()\n",
    "    elif features==2:\n",
    "        score=[np.round(model_Sp.predict_proba([encoded_TR[0][:1]])[:,1],decimals=2).item(),np.round(model_Avd_Sp.predict_proba([encoded_TR[0][1:2]])[:,1],decimals=2).item()]\n",
    "    return score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4ff42230-8c48-44e9-9222-bd3c879724b5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TR bad score = 0.23\n",
      "TR good score = 0.84\n"
     ]
    }
   ],
   "source": [
    "TR_bad='TTAGCGAATGGCGAAATTCGTAAACGCCCTCTGATCGAAACCAACGGCGAAACGGGTGAGATCGTGTGGG'\n",
    "print('TR bad score =',score(TR_bad))\n",
    "TR_good='AAATGATCGCCAAATCTGAACAGGAAATTGGCAAAGCAACCGCTAAATACTTTTTCTACTCAAACATTAT'\n",
    "print('TR good score =',score(TR_good))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e8d613b2-9373-4929-b878-ee66e6e1e3d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# | export\n",
    "def score_list(TR_seq_list:list, #A list of strings of TRs DNA sequences\n",
    "TR_name_list:list, #A list of strings of TRs names\n",
    "features=1 #The number of features to use\n",
    "              ):\n",
    "    \"\"\"Calculates the score for every TR in the list and returns them in a dataframe format. If features=2, returns the score according to each feature (better to have both high).\"\"\"\n",
    "\n",
    "    encoded_TR=encoding.encode_tr_list(TR_seq_list,features)\n",
    "    if features==1:\n",
    "        score=np.round(model_Sp.predict_proba(encoded_TR)[:,1],decimals=2)\n",
    "        score_df=pd.DataFrame({\n",
    "            'TR_Name':TR_name_list,\n",
    "            'TR_Seq':TR_seq_list,\n",
    "            'TR_Score':score\n",
    "        })\n",
    "        return score_df\n",
    "    else:\n",
    "        score_Sp=np.round(model_Sp.predict_proba(encoded_TR[:,:1])[:,1],decimals=2)\n",
    "        score_Avd_Sp=np.round(model_Avd_Sp.predict_proba(encoded_TR[:,1:2])[:,1],decimals=2)\n",
    "        score_df=pd.DataFrame({\n",
    "            'TR_Name':TR_name_list,\n",
    "            'TR_Seq':TR_seq_list,\n",
    "            'TR_Score_Sp':score_Sp,\n",
    "            'TR_Score_Avd':score_Avd_Sp\n",
    "        })\n",
    "        return score_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f1496834-ef20-429b-8307-64139f42dd00",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TR_Name</th>\n",
       "      <th>TR_Seq</th>\n",
       "      <th>TR_Score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TR_bad_1</td>\n",
       "      <td>TTAGCGAATGGCGAAATTCGTAAACGCCCTCTGATCGAAACCAACG...</td>\n",
       "      <td>0.23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TR_bad_2</td>\n",
       "      <td>AAACGCCCTCTGATCGAAACCAACGGCGAAACGGGTGAGATCGTGT...</td>\n",
       "      <td>0.05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TR_bad_3</td>\n",
       "      <td>GGTTTCTCTAAGGAGTCCATTCTGCCGAAGCGCAACTCCGACAAGC...</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TR_bad_4</td>\n",
       "      <td>CAAGCTGATCGCGCGTAAGAAGGACTGGGATCCGAAGAAGTACGGT...</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TR_bad_5</td>\n",
       "      <td>ACCCGATTGACTTCCTCGAGGCGAAGGGGTACAAGGAGGTGAAGAA...</td>\n",
       "      <td>0.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>TR_bad_6</td>\n",
       "      <td>AGTACTCCCTGTTCGAGCTGGAGAATGGTCGTAAGCGTATGCTGGC...</td>\n",
       "      <td>0.08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>TR_bad_7</td>\n",
       "      <td>CAGCACAAGCACTACCTGGACGAGATTATTGAGCAGATTTCTGAGT...</td>\n",
       "      <td>0.06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>TR_bad_8</td>\n",
       "      <td>ACGCGAATCTGGATAAGGTCCTGTCTGCCTACAATAAGCACCGTGA...</td>\n",
       "      <td>0.12</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    TR_Name                                             TR_Seq  TR_Score\n",
       "0  TR_bad_1  TTAGCGAATGGCGAAATTCGTAAACGCCCTCTGATCGAAACCAACG...      0.23\n",
       "1  TR_bad_2  AAACGCCCTCTGATCGAAACCAACGGCGAAACGGGTGAGATCGTGT...      0.05\n",
       "2  TR_bad_3  GGTTTCTCTAAGGAGTCCATTCTGCCGAAGCGCAACTCCGACAAGC...      0.00\n",
       "3  TR_bad_4  CAAGCTGATCGCGCGTAAGAAGGACTGGGATCCGAAGAAGTACGGT...      0.00\n",
       "4  TR_bad_5  ACCCGATTGACTTCCTCGAGGCGAAGGGGTACAAGGAGGTGAAGAA...      0.01\n",
       "5  TR_bad_6  AGTACTCCCTGTTCGAGCTGGAGAATGGTCGTAAGCGTATGCTGGC...      0.08\n",
       "6  TR_bad_7  CAGCACAAGCACTACCTGGACGAGATTATTGAGCAGATTTCTGAGT...      0.06\n",
       "7  TR_bad_8  ACGCGAATCTGGATAAGGTCCTGTCTGCCTACAATAAGCACCGTGA...      0.12"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TR_bad=[\n",
    "     'TTAGCGAATGGCGAAATTCGTAAACGCCCTCTGATCGAAACCAACGGCGAAACGGGTGAGATCGTGTGGG',\n",
    "     'AAACGCCCTCTGATCGAAACCAACGGCGAAACGGGTGAGATCGTGTGGGACAAAGGTCGTGATTTCGCTA',\n",
    "    'GGTTTCTCTAAGGAGTCCATTCTGCCGAAGCGCAACTCCGACAAGCTGATCGCGCGTAAGAAGGACTGGG',\n",
    "     'CAAGCTGATCGCGCGTAAGAAGGACTGGGATCCGAAGAAGTACGGTGGCTTCGATTCTCCGACCGTGGCG',\n",
    "     'ACCCGATTGACTTCCTCGAGGCGAAGGGGTACAAGGAGGTGAAGAAGGATCTGATTATCAAGCTGCCGAA',\n",
    "     'AGTACTCCCTGTTCGAGCTGGAGAATGGTCGTAAGCGTATGCTGGCGTCTGCGGGTGAGCTGCAGAAGGG',\n",
    "     'CAGCACAAGCACTACCTGGACGAGATTATTGAGCAGATTTCTGAGTTTTCTAAGCGCGTGATTCTGGCGG',\n",
    "     'ACGCGAATCTGGATAAGGTCCTGTCTGCCTACAATAAGCACCGTGATAAGCCGATCCGTGAGCAGGCGGA',   \n",
    " ]\n",
    "\n",
    "score_list(TR_bad,['TR_bad_'+str(k) for k in range (1,9)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fac0795f-6041-4636-b7f5-0e92a175c2d6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TR_Name</th>\n",
       "      <th>TR_Seq</th>\n",
       "      <th>TR_Score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TR_good_1</td>\n",
       "      <td>AAATGATCGCCAAATCTGAACAGGAAATTGGCAAAGCAACCGCTAA...</td>\n",
       "      <td>0.84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TR_good_2</td>\n",
       "      <td>TCAAACATTATGAATTTCTTCAAAACCGAAATCACCTTAGCGAATG...</td>\n",
       "      <td>0.82</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TR_good_3</td>\n",
       "      <td>ATGCCTCAAGTAAACATCGTTAAAAAGACTGAGGTGCAGACTGGCG...</td>\n",
       "      <td>0.76</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TR_good_4</td>\n",
       "      <td>GGATCCGAAGAAGTACGGTGGCTTCGATTCTCCGACCGTGGCGTAC...</td>\n",
       "      <td>0.74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TR_good_5</td>\n",
       "      <td>AGCGTATGCTGGCGTCTGCGGGTGAGCTGCAGAAGGGGAACGAGTT...</td>\n",
       "      <td>0.83</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>TR_good_6</td>\n",
       "      <td>GCAGAAGGGGAACGAGTTGGCCCTTCCGTCCAAGTACGTGAACTTC...</td>\n",
       "      <td>0.55</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>TR_good_7</td>\n",
       "      <td>CAGAAGCAGCTGTTCGTGGAGCAGCACAAGCACTACCTGGACGAGA...</td>\n",
       "      <td>0.81</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>TR_good_8</td>\n",
       "      <td>CTAAGCGCGTGATTCTGGCGGACGCGAATCTGGATAAGGTCCTGTC...</td>\n",
       "      <td>0.81</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     TR_Name                                             TR_Seq  TR_Score\n",
       "0  TR_good_1  AAATGATCGCCAAATCTGAACAGGAAATTGGCAAAGCAACCGCTAA...      0.84\n",
       "1  TR_good_2  TCAAACATTATGAATTTCTTCAAAACCGAAATCACCTTAGCGAATG...      0.82\n",
       "2  TR_good_3  ATGCCTCAAGTAAACATCGTTAAAAAGACTGAGGTGCAGACTGGCG...      0.76\n",
       "3  TR_good_4  GGATCCGAAGAAGTACGGTGGCTTCGATTCTCCGACCGTGGCGTAC...      0.74\n",
       "4  TR_good_5  AGCGTATGCTGGCGTCTGCGGGTGAGCTGCAGAAGGGGAACGAGTT...      0.83\n",
       "5  TR_good_6  GCAGAAGGGGAACGAGTTGGCCCTTCCGTCCAAGTACGTGAACTTC...      0.55\n",
       "6  TR_good_7  CAGAAGCAGCTGTTCGTGGAGCAGCACAAGCACTACCTGGACGAGA...      0.81\n",
       "7  TR_good_8  CTAAGCGCGTGATTCTGGCGGACGCGAATCTGGATAAGGTCCTGTC...      0.81"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TR_good=[\n",
    "     'AAATGATCGCCAAATCTGAACAGGAAATTGGCAAAGCAACCGCTAAATACTTTTTCTACTCAAACATTAT',\n",
    "     'TCAAACATTATGAATTTCTTCAAAACCGAAATCACCTTAGCGAATGGCGAAATTCGTAAACGCCCTCTGA',\n",
    "     'ATGCCTCAAGTAAACATCGTTAAAAAGACTGAGGTGCAGACTGGCGGTTTCTCTAAGGAGTCCATTCTGC',\n",
    "     'GGATCCGAAGAAGTACGGTGGCTTCGATTCTCCGACCGTGGCGTACTCTGTTCTGGTGGTCGCCAAGGTC',\n",
    "     'AGCGTATGCTGGCGTCTGCGGGTGAGCTGCAGAAGGGGAACGAGTTGGCCCTTCCGTCCAAGTACGTGAA',\n",
    "     'GCAGAAGGGGAACGAGTTGGCCCTTCCGTCCAAGTACGTGAACTTCCTGTACCTGGCCTCGCACTACGAG',\n",
    "     'CAGAAGCAGCTGTTCGTGGAGCAGCACAAGCACTACCTGGACGAGATTATTGAGCAGATTTCTGAGTTTT',\n",
    "     'CTAAGCGCGTGATTCTGGCGGACGCGAATCTGGATAAGGTCCTGTCTGCCTACAATAAGCACCGTGATAA'\n",
    "     ]\n",
    "\n",
    "score_list(TR_good,['TR_good_'+str(k) for k in range (1,9)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a96dd4c7-1556-4c30-8c65-34593c93f3bc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TR_Name</th>\n",
       "      <th>TR_Seq</th>\n",
       "      <th>TR_Score_Sp</th>\n",
       "      <th>TR_Score_Avd</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TR_bad_1</td>\n",
       "      <td>TTAGCGAATGGCGAAATTCGTAAACGCCCTCTGATCGAAACCAACG...</td>\n",
       "      <td>0.23</td>\n",
       "      <td>0.63</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TR_bad_2</td>\n",
       "      <td>AAACGCCCTCTGATCGAAACCAACGGCGAAACGGGTGAGATCGTGT...</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TR_bad_3</td>\n",
       "      <td>GGTTTCTCTAAGGAGTCCATTCTGCCGAAGCGCAACTCCGACAAGC...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TR_bad_4</td>\n",
       "      <td>CAAGCTGATCGCGCGTAAGAAGGACTGGGATCCGAAGAAGTACGGT...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.51</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TR_bad_5</td>\n",
       "      <td>ACCCGATTGACTTCCTCGAGGCGAAGGGGTACAAGGAGGTGAAGAA...</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.59</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>TR_bad_6</td>\n",
       "      <td>AGTACTCCCTGTTCGAGCTGGAGAATGGTCGTAAGCGTATGCTGGC...</td>\n",
       "      <td>0.08</td>\n",
       "      <td>0.54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>TR_bad_7</td>\n",
       "      <td>CAGCACAAGCACTACCTGGACGAGATTATTGAGCAGATTTCTGAGT...</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>TR_bad_8</td>\n",
       "      <td>ACGCGAATCTGGATAAGGTCCTGTCTGCCTACAATAAGCACCGTGA...</td>\n",
       "      <td>0.12</td>\n",
       "      <td>0.04</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    TR_Name                                             TR_Seq  TR_Score_Sp  \\\n",
       "0  TR_bad_1  TTAGCGAATGGCGAAATTCGTAAACGCCCTCTGATCGAAACCAACG...         0.23   \n",
       "1  TR_bad_2  AAACGCCCTCTGATCGAAACCAACGGCGAAACGGGTGAGATCGTGT...         0.05   \n",
       "2  TR_bad_3  GGTTTCTCTAAGGAGTCCATTCTGCCGAAGCGCAACTCCGACAAGC...         0.00   \n",
       "3  TR_bad_4  CAAGCTGATCGCGCGTAAGAAGGACTGGGATCCGAAGAAGTACGGT...         0.00   \n",
       "4  TR_bad_5  ACCCGATTGACTTCCTCGAGGCGAAGGGGTACAAGGAGGTGAAGAA...         0.01   \n",
       "5  TR_bad_6  AGTACTCCCTGTTCGAGCTGGAGAATGGTCGTAAGCGTATGCTGGC...         0.08   \n",
       "6  TR_bad_7  CAGCACAAGCACTACCTGGACGAGATTATTGAGCAGATTTCTGAGT...         0.06   \n",
       "7  TR_bad_8  ACGCGAATCTGGATAAGGTCCTGTCTGCCTACAATAAGCACCGTGA...         0.12   \n",
       "\n",
       "   TR_Score_Avd  \n",
       "0          0.63  \n",
       "1          0.42  \n",
       "2          0.30  \n",
       "3          0.51  \n",
       "4          0.59  \n",
       "5          0.54  \n",
       "6          0.29  \n",
       "7          0.04  "
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TR_bad=[\n",
    "     'TTAGCGAATGGCGAAATTCGTAAACGCCCTCTGATCGAAACCAACGGCGAAACGGGTGAGATCGTGTGGG',\n",
    "     'AAACGCCCTCTGATCGAAACCAACGGCGAAACGGGTGAGATCGTGTGGGACAAAGGTCGTGATTTCGCTA',\n",
    "    'GGTTTCTCTAAGGAGTCCATTCTGCCGAAGCGCAACTCCGACAAGCTGATCGCGCGTAAGAAGGACTGGG',\n",
    "     'CAAGCTGATCGCGCGTAAGAAGGACTGGGATCCGAAGAAGTACGGTGGCTTCGATTCTCCGACCGTGGCG',\n",
    "     'ACCCGATTGACTTCCTCGAGGCGAAGGGGTACAAGGAGGTGAAGAAGGATCTGATTATCAAGCTGCCGAA',\n",
    "     'AGTACTCCCTGTTCGAGCTGGAGAATGGTCGTAAGCGTATGCTGGCGTCTGCGGGTGAGCTGCAGAAGGG',\n",
    "     'CAGCACAAGCACTACCTGGACGAGATTATTGAGCAGATTTCTGAGTTTTCTAAGCGCGTGATTCTGGCGG',\n",
    "     'ACGCGAATCTGGATAAGGTCCTGTCTGCCTACAATAAGCACCGTGATAAGCCGATCCGTGAGCAGGCGGA',   \n",
    " ]\n",
    "\n",
    "score_list(TR_bad,['TR_bad_'+str(k) for k in range (1,9)],2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4d24739a-c578-45d4-9893-0c4d2670493d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TR_Name</th>\n",
       "      <th>TR_Seq</th>\n",
       "      <th>TR_Score_Sp</th>\n",
       "      <th>TR_Score_Avd</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TR_good_1</td>\n",
       "      <td>AAATGATCGCCAAATCTGAACAGGAAATTGGCAAAGCAACCGCTAA...</td>\n",
       "      <td>0.84</td>\n",
       "      <td>0.80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TR_good_2</td>\n",
       "      <td>TCAAACATTATGAATTTCTTCAAAACCGAAATCACCTTAGCGAATG...</td>\n",
       "      <td>0.82</td>\n",
       "      <td>0.78</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TR_good_3</td>\n",
       "      <td>ATGCCTCAAGTAAACATCGTTAAAAAGACTGAGGTGCAGACTGGCG...</td>\n",
       "      <td>0.76</td>\n",
       "      <td>0.84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TR_good_4</td>\n",
       "      <td>GGATCCGAAGAAGTACGGTGGCTTCGATTCTCCGACCGTGGCGTAC...</td>\n",
       "      <td>0.74</td>\n",
       "      <td>0.75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TR_good_5</td>\n",
       "      <td>AGCGTATGCTGGCGTCTGCGGGTGAGCTGCAGAAGGGGAACGAGTT...</td>\n",
       "      <td>0.83</td>\n",
       "      <td>0.88</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>TR_good_6</td>\n",
       "      <td>GCAGAAGGGGAACGAGTTGGCCCTTCCGTCCAAGTACGTGAACTTC...</td>\n",
       "      <td>0.55</td>\n",
       "      <td>0.58</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>TR_good_7</td>\n",
       "      <td>CAGAAGCAGCTGTTCGTGGAGCAGCACAAGCACTACCTGGACGAGA...</td>\n",
       "      <td>0.81</td>\n",
       "      <td>0.34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>TR_good_8</td>\n",
       "      <td>CTAAGCGCGTGATTCTGGCGGACGCGAATCTGGATAAGGTCCTGTC...</td>\n",
       "      <td>0.81</td>\n",
       "      <td>0.82</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     TR_Name                                             TR_Seq  TR_Score_Sp  \\\n",
       "0  TR_good_1  AAATGATCGCCAAATCTGAACAGGAAATTGGCAAAGCAACCGCTAA...         0.84   \n",
       "1  TR_good_2  TCAAACATTATGAATTTCTTCAAAACCGAAATCACCTTAGCGAATG...         0.82   \n",
       "2  TR_good_3  ATGCCTCAAGTAAACATCGTTAAAAAGACTGAGGTGCAGACTGGCG...         0.76   \n",
       "3  TR_good_4  GGATCCGAAGAAGTACGGTGGCTTCGATTCTCCGACCGTGGCGTAC...         0.74   \n",
       "4  TR_good_5  AGCGTATGCTGGCGTCTGCGGGTGAGCTGCAGAAGGGGAACGAGTT...         0.83   \n",
       "5  TR_good_6  GCAGAAGGGGAACGAGTTGGCCCTTCCGTCCAAGTACGTGAACTTC...         0.55   \n",
       "6  TR_good_7  CAGAAGCAGCTGTTCGTGGAGCAGCACAAGCACTACCTGGACGAGA...         0.81   \n",
       "7  TR_good_8  CTAAGCGCGTGATTCTGGCGGACGCGAATCTGGATAAGGTCCTGTC...         0.81   \n",
       "\n",
       "   TR_Score_Avd  \n",
       "0          0.80  \n",
       "1          0.78  \n",
       "2          0.84  \n",
       "3          0.75  \n",
       "4          0.88  \n",
       "5          0.58  \n",
       "6          0.34  \n",
       "7          0.82  "
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TR_good=[\n",
    "     'AAATGATCGCCAAATCTGAACAGGAAATTGGCAAAGCAACCGCTAAATACTTTTTCTACTCAAACATTAT',\n",
    "     'TCAAACATTATGAATTTCTTCAAAACCGAAATCACCTTAGCGAATGGCGAAATTCGTAAACGCCCTCTGA',\n",
    "     'ATGCCTCAAGTAAACATCGTTAAAAAGACTGAGGTGCAGACTGGCGGTTTCTCTAAGGAGTCCATTCTGC',\n",
    "     'GGATCCGAAGAAGTACGGTGGCTTCGATTCTCCGACCGTGGCGTACTCTGTTCTGGTGGTCGCCAAGGTC',\n",
    "     'AGCGTATGCTGGCGTCTGCGGGTGAGCTGCAGAAGGGGAACGAGTTGGCCCTTCCGTCCAAGTACGTGAA',\n",
    "     'GCAGAAGGGGAACGAGTTGGCCCTTCCGTCCAAGTACGTGAACTTCCTGTACCTGGCCTCGCACTACGAG',\n",
    "     'CAGAAGCAGCTGTTCGTGGAGCAGCACAAGCACTACCTGGACGAGATTATTGAGCAGATTTCTGAGTTTT',\n",
    "     'CTAAGCGCGTGATTCTGGCGGACGCGAATCTGGATAAGGTCCTGTCTGCCTACAATAAGCACCGTGATAA'\n",
    "     ]\n",
    "\n",
    "score_list(TR_good,['TR_good_'+str(k) for k in range (1,9)],2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8156475d-37c1-463e-bb4c-4e80d45daed0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# | export\n",
    "def DGR_percentage(TR_seq:str #A string of the TR DNA sequence\n",
    "         ):\n",
    "\n",
    "    \"\"\"Calculates the predicted DGR mutagenesis percentage of a given TR sequence (100 = perfect TR and 0 = crappy TR)\"\"\"\n",
    "    encoded_TR=encoding.encode_tr_list([TR_seq],2)\n",
    "    rate=model_whole.predict(encoded_TR)[0]\n",
    "    return 10**rate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "35ae5c70-2dea-4651-8bd0-2366f2eb9444",
   "metadata": {},
   "outputs": [],
   "source": [
    "# | export\n",
    "def DGR_percentage_list(TR_seq_list:list, #A list of strings of TRs DNA sequences\n",
    "TR_name_list:list, #A list of strings of TRs names\n",
    "              ):\n",
    "    \"\"\"Calculates the predicted DGR mutagenesis percentage for every TR in the list and returns them in a dataframe format\"\"\"\n",
    "\n",
    "    encoded_TR=encoding.encode_tr_list(TR_seq_list,2)\n",
    "    rates=model_whole.predict(encoded_TR)\n",
    "    score_df=pd.DataFrame({\n",
    "        'TR_Name':TR_name_list,\n",
    "        'TR_Seq':10**rates\n",
    "    })\n",
    "    return score_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "723fb241-75a5-46ee-9193-c53d9e0aba92",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TR_Name</th>\n",
       "      <th>TR_Seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TR_bad_1</td>\n",
       "      <td>0.427049</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TR_bad_2</td>\n",
       "      <td>0.110769</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TR_bad_3</td>\n",
       "      <td>0.019986</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TR_bad_4</td>\n",
       "      <td>0.025612</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TR_bad_5</td>\n",
       "      <td>0.045752</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>TR_bad_6</td>\n",
       "      <td>0.172833</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>TR_bad_7</td>\n",
       "      <td>0.092111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>TR_bad_8</td>\n",
       "      <td>0.039965</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    TR_Name    TR_Seq\n",
       "0  TR_bad_1  0.427049\n",
       "1  TR_bad_2  0.110769\n",
       "2  TR_bad_3  0.019986\n",
       "3  TR_bad_4  0.025612\n",
       "4  TR_bad_5  0.045752\n",
       "5  TR_bad_6  0.172833\n",
       "6  TR_bad_7  0.092111\n",
       "7  TR_bad_8  0.039965"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TR_bad=[\n",
    "     'TTAGCGAATGGCGAAATTCGTAAACGCCCTCTGATCGAAACCAACGGCGAAACGGGTGAGATCGTGTGGG',\n",
    "     'AAACGCCCTCTGATCGAAACCAACGGCGAAACGGGTGAGATCGTGTGGGACAAAGGTCGTGATTTCGCTA',\n",
    "    'GGTTTCTCTAAGGAGTCCATTCTGCCGAAGCGCAACTCCGACAAGCTGATCGCGCGTAAGAAGGACTGGG',\n",
    "     'CAAGCTGATCGCGCGTAAGAAGGACTGGGATCCGAAGAAGTACGGTGGCTTCGATTCTCCGACCGTGGCG',\n",
    "     'ACCCGATTGACTTCCTCGAGGCGAAGGGGTACAAGGAGGTGAAGAAGGATCTGATTATCAAGCTGCCGAA',\n",
    "     'AGTACTCCCTGTTCGAGCTGGAGAATGGTCGTAAGCGTATGCTGGCGTCTGCGGGTGAGCTGCAGAAGGG',\n",
    "     'CAGCACAAGCACTACCTGGACGAGATTATTGAGCAGATTTCTGAGTTTTCTAAGCGCGTGATTCTGGCGG',\n",
    "     'ACGCGAATCTGGATAAGGTCCTGTCTGCCTACAATAAGCACCGTGATAAGCCGATCCGTGAGCAGGCGGA',   \n",
    " ]\n",
    "\n",
    "DGR_percentage_list(TR_bad,['TR_bad_'+str(k) for k in range (1,9)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "42fb7d07-a07a-4196-be69-0a0adf1c54eb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TR_Name</th>\n",
       "      <th>TR_Seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TR_good_1</td>\n",
       "      <td>3.067841</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TR_good_2</td>\n",
       "      <td>2.721600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TR_good_3</td>\n",
       "      <td>2.808427</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TR_good_4</td>\n",
       "      <td>1.919474</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TR_good_5</td>\n",
       "      <td>4.142115</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>TR_good_6</td>\n",
       "      <td>0.811313</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>TR_good_7</td>\n",
       "      <td>0.940305</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>TR_good_8</td>\n",
       "      <td>3.040041</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     TR_Name    TR_Seq\n",
       "0  TR_good_1  3.067841\n",
       "1  TR_good_2  2.721600\n",
       "2  TR_good_3  2.808427\n",
       "3  TR_good_4  1.919474\n",
       "4  TR_good_5  4.142115\n",
       "5  TR_good_6  0.811313\n",
       "6  TR_good_7  0.940305\n",
       "7  TR_good_8  3.040041"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TR_good=[\n",
    "     'AAATGATCGCCAAATCTGAACAGGAAATTGGCAAAGCAACCGCTAAATACTTTTTCTACTCAAACATTAT',\n",
    "     'TCAAACATTATGAATTTCTTCAAAACCGAAATCACCTTAGCGAATGGCGAAATTCGTAAACGCCCTCTGA',\n",
    "     'ATGCCTCAAGTAAACATCGTTAAAAAGACTGAGGTGCAGACTGGCGGTTTCTCTAAGGAGTCCATTCTGC',\n",
    "     'GGATCCGAAGAAGTACGGTGGCTTCGATTCTCCGACCGTGGCGTACTCTGTTCTGGTGGTCGCCAAGGTC',\n",
    "     'AGCGTATGCTGGCGTCTGCGGGTGAGCTGCAGAAGGGGAACGAGTTGGCCCTTCCGTCCAAGTACGTGAA',\n",
    "     'GCAGAAGGGGAACGAGTTGGCCCTTCCGTCCAAGTACGTGAACTTCCTGTACCTGGCCTCGCACTACGAG',\n",
    "     'CAGAAGCAGCTGTTCGTGGAGCAGCACAAGCACTACCTGGACGAGATTATTGAGCAGATTTCTGAGTTTT',\n",
    "     'CTAAGCGCGTGATTCTGGCGGACGCGAATCTGGATAAGGTCCTGTCTGCCTACAATAAGCACCGTGATAA'\n",
    "     ]\n",
    "\n",
    "DGR_percentage_list(TR_good,['TR_good_'+str(k) for k in range (1,9)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c1cc23f-b21b-41a2-b03f-627eec8b5428",
   "metadata": {},
   "outputs": [],
   "source": [
    "# | export\n",
    "\n",
    "codon_usage_ecoli = {\n",
    "    \"F\": {\"TTT\": 0.57, \"TTC\": 0.43},\n",
    "    \"L\": {\"TTA\": 0.15, \"TTG\": 0.12, \"CTT\": 0.12, \"CTC\": 0.10, \"CTA\": 0.05, \"CTG\": 0.46},\n",
    "    \"S\": {\"TCT\": 0.11, \"TCC\": 0.11, \"TCA\": 0.15, \"TCG\": 0.16, \"AGT\": 0.14, \"AGC\": 0.33},\n",
    "    \"Y\": {\"TAT\": 0.53, \"TAC\": 0.47},\n",
    "    \"*\": {\"TAA\": 0.64, \"TAG\": 0.00, \"TGA\": 0.36},\n",
    "    \"C\": {\"TGT\": 0.42, \"TGC\": 0.58},\n",
    "    \"W\": {\"TGG\": 1.00},\n",
    "    \"P\": {\"CCT\": 0.17, \"CCC\": 0.13, \"CCA\": 0.14, \"CCG\": 0.55},\n",
    "    \"H\": {\"CAT\": 0.55, \"CAC\": 0.45},\n",
    "    \"Q\": {\"CAA\": 0.30, \"CAG\": 0.70},\n",
    "    \"R\": {\"CGT\": 0.36, \"CGC\": 0.44, \"CGA\": 0.07, \"CGG\": 0.07, \"AGA\": 0.07, \"AGG\": 0.00},\n",
    "    \"I\": {\"ATT\": 0.58, \"ATC\": 0.35, \"ATA\": 0.07},\n",
    "    \"M\": {\"ATG\": 1.00},\n",
    "    \"T\": {\"ACT\": 0.16, \"ACC\": 0.47, \"ACA\": 0.13, \"ACG\": 0.24},\n",
    "    \"N\": {\"AAT\": 0.47, \"AAC\": 0.53},\n",
    "    \"K\": {\"AAA\": 0.73, \"AAG\": 0.27},\n",
    "    \"V\": {\"GTT\": 0.25, \"GTC\": 0.18, \"GTA\": 0.17, \"GTG\": 0.40},\n",
    "    \"A\": {\"GCT\": 0.11, \"GCC\": 0.31, \"GCA\": 0.20, \"GCG\": 0.38},\n",
    "    \"D\": {\"GAT\": 0.65, \"GAC\": 0.35},\n",
    "    \"E\": {\"GAA\": 0.70, \"GAG\": 0.30},\n",
    "    \"G\": {\"GGT\": 0.29, \"GGC\": 0.46, \"GGA\": 0.13, \"GGG\": 0.12}\n",
    "}\n",
    "\n",
    "# Genetic code: codon -> amino acid\n",
    "genetic_code = {}\n",
    "for aa, codons in codon_usage_ecoli.items():\n",
    "    for codon in codons:\n",
    "        genetic_code[codon] = aa\n",
    "\n",
    "CODON_TO_AA = {\n",
    "    'ATA': 'I', 'ATC': 'I', 'ATT': 'I', 'ATG': 'M',\n",
    "    'ACA': 'T', 'ACC': 'T', 'ACG': 'T', 'ACT': 'T',\n",
    "    'AAC': 'N', 'AAT': 'N', 'AAA': 'K', 'AAG': 'K',\n",
    "    'AGC': 'S', 'AGT': 'S', 'AGA': 'R', 'AGG': 'R',\n",
    "    'CTA': 'L', 'CTC': 'L', 'CTG': 'L', 'CTT': 'L',\n",
    "    'CCA': 'P', 'CCC': 'P', 'CCG': 'P', 'CCT': 'P',\n",
    "    'CAC': 'H', 'CAT': 'H', 'CAA': 'Q', 'CAG': 'Q',\n",
    "    'CGA': 'R', 'CGC': 'R', 'CGG': 'R', 'CGT': 'R',\n",
    "    'GTA': 'V', 'GTC': 'V', 'GTG': 'V', 'GTT': 'V',\n",
    "    'GCA': 'A', 'GCC': 'A', 'GCG': 'A', 'GCT': 'A',\n",
    "    'GAC': 'D', 'GAT': 'D', 'GAA': 'E', 'GAG': 'E',\n",
    "    'GGA': 'G', 'GGC': 'G', 'GGG': 'G', 'GGT': 'G',\n",
    "    'TCA': 'S', 'TCC': 'S', 'TCG': 'S', 'TCT': 'S',\n",
    "    'TTC': 'F', 'TTT': 'F', 'TTA': 'L', 'TTG': 'L',\n",
    "    'TAC': 'Y', 'TAT': 'Y', 'TAA': '*', 'TAG': '*',\n",
    "    'TGC': 'C', 'TGT': 'C', 'TGA': '*', 'TGG': 'W'\n",
    "}\n",
    "\n",
    "\n",
    "AA = '-ACDEFGHIKLMNPQRSTVWY'\n",
    "AA_IDS = {k: i for i, k in enumerate(AA)}\n",
    "\n",
    "\n",
    "AA_TO_CODONS = {\n",
    "    'A': ['GCT', 'GCC', 'GCA', 'GCG'],  # Alanine\n",
    "    'C': ['TGT', 'TGC'],               # Cysteine\n",
    "    'D': ['GAT', 'GAC'],               # Aspartic acid\n",
    "    'E': ['GAA', 'GAG'],               # Glutamic acid\n",
    "    'F': ['TTT', 'TTC'],               # Phenylalanine\n",
    "    'G': ['GGT', 'GGC', 'GGA', 'GGG'], # Glycine\n",
    "    'H': ['CAT', 'CAC'],               # Histidine\n",
    "    'I': ['ATT', 'ATC', 'ATA'],        # Isoleucine\n",
    "    'K': ['AAA', 'AAG'],               # Lysine\n",
    "    'L': ['TTA', 'TTG', 'CTT', 'CTC', 'CTA', 'CTG'],  # Leucine\n",
    "    'M': ['ATG'],                      # Methionine (Start codon)\n",
    "    'N': ['AAT', 'AAC'],               # Asparagine\n",
    "    'P': ['CCT', 'CCC', 'CCA', 'CCG'], # Proline\n",
    "    'Q': ['CAA', 'CAG'],               # Glutamine\n",
    "    'R': ['CGT', 'CGC', 'CGA', 'CGG', 'AGA', 'AGG'],  # Arginine\n",
    "    'S': ['TCT', 'TCC', 'TCA', 'TCG', 'AGT', 'AGC'],  # Serine\n",
    "    'T': ['ACT', 'ACC', 'ACA', 'ACG'], # Threonine\n",
    "    'V': ['GTT', 'GTC', 'GTA', 'GTG'], # Valine\n",
    "    'W': ['TGG'],                      # Tryptophan\n",
    "    'Y': ['TAT', 'TAC'],               # Tyrosine\n",
    "    '*': ['TAA', 'TAG', 'TGA'],        # Stop codons\n",
    "    '-': []                            # Gap (no codons)\n",
    "}\n",
    "\n",
    "def nt_to_codons(nt_sequence):\n",
    "    \"\"\"\n",
    "    Converts a nucleotide sequence into a list of codons.\n",
    "    \n",
    "    Args:\n",
    "        nt_sequence (str): A string of nucleotides (e.g., \"ATGGCATGA\").\n",
    "        \n",
    "    Returns:\n",
    "        list: A list of codons (e.g., ['ATG', 'GCA', 'TGA']).\n",
    "    \"\"\"\n",
    "    # Ensure the sequence length is a multiple of 3 by trimming extra nucleotides\n",
    "    trimmed_sequence = nt_sequence[:len(nt_sequence) // 3 * 3]\n",
    "    \n",
    "    # Split the sequence into codons (groups of 3 nucleotides)\n",
    "    codons = [trimmed_sequence[i:i+3] for i in range(0, len(trimmed_sequence), 3)]\n",
    "    \n",
    "    return codons\n",
    "\n",
    "def codons_to_aa(codons):\n",
    "    \"\"\"\n",
    "    Converts a list of codons into a sequence of amino acids.\n",
    "    \n",
    "    Args:\n",
    "        codons (list): A list of codon strings (e.g., ['ATG', 'GCA', 'TGA']).\n",
    "        \n",
    "    Returns:\n",
    "        str: The translated amino acid sequence (e.g., 'M*A').\n",
    "    \"\"\"\n",
    "    aa_sequence = \"\"\n",
    "    for codon in codons:\n",
    "        # Convert codon to amino acid; default to 'X' for invalid codons\n",
    "        aa_sequence += CODON_TO_AA.get(codon, 'X')\n",
    "    return aa_sequence\n",
    "\n",
    "def nt_prot(nt):\n",
    "    return codons_to_aa(nt_to_codons(nt))\n",
    "\n",
    "def pareto_front(sequences):\n",
    "    \"\"\"\n",
    "    Keep only Pareto-optimal sequences:\n",
    "    no other sequence has both scores >= and at least one strictly greater.\n",
    "    \"\"\"\n",
    "    pareto = []\n",
    "    for s in sequences:\n",
    "        dominated = False\n",
    "        for t in sequences:\n",
    "            if (t[\"Score_TRSp\"] >= s[\"Score_TRSp\"] and\n",
    "                t[\"Score_TRSpAvd\"] >= s[\"Score_TRSpAvd\"] and\n",
    "                (t[\"Score_TRSp\"] > s[\"Score_TRSp\"] or\n",
    "                 t[\"Score_TRSpAvd\"] > s[\"Score_TRSpAvd\"])):\n",
    "                dominated = True\n",
    "                break\n",
    "        if not dominated:\n",
    "            pareto.append(s)\n",
    "    return pareto\n",
    "\n",
    "def propose_single_codon_changes(seq, frame_offset=0,rank_max=2,forbidden_positions=[],codon_usage=codon_usage_ecoli):\n",
    "    \"\"\"\n",
    "    Generate all single-codon synonymous changes from a given sequence.\n",
    "    Takes into account reading frame offset (Position % 3).\n",
    "    Returns list of (new_seq, new_changes).\n",
    "    \"\"\"\n",
    "    variants = []\n",
    "    for i in range(frame_offset, len(seq)-2, 3):\n",
    "      if not( (i in forbidden_positions) or (i+1 in forbidden_positions) or (i+2 in forbidden_positions)): \n",
    "        codon = seq[i:i+3]\n",
    "        aa = genetic_code.get(codon, \"?\")\n",
    "        if aa not in codon_usage:\n",
    "            continue\n",
    "\n",
    "        ranked = sorted(codon_usage[aa].items(), key=lambda x: x[1], reverse=True)\n",
    "        for new_codon, usage in ranked[:rank_max]:\n",
    "            if new_codon == codon:\n",
    "                continue\n",
    "            new_seq = seq[:i] + new_codon + seq[i+3:]\n",
    "            \n",
    "            variants.append(new_seq)\n",
    "    return variants\n",
    "\n",
    "def evaluate_sequences(variants):\n",
    "    \"\"\"\n",
    "    Encode sequences, run classifiers, return list of dicts with scores.\n",
    "    \"\"\"\n",
    "    seqs = [v for v in variants]\n",
    "    Scores=encoding.encode_tr_list(seqs,2)\n",
    "    Score_TRSp = [a for [a,b] in Scores]\n",
    "    Score_TRSpAvd = [b for [a,b] in Scores]\n",
    "    plt.figure(figsize=(6,6))\n",
    "    plt.scatter(Score_TRSp, Score_TRSpAvd, alpha=0.7)\n",
    "    plt.xlabel(\"Score_TRSp\")\n",
    "    plt.ylabel(\"Score_TRSpAvd\")\n",
    "    plt.grid(True)\n",
    "    plt.show()\n",
    "    results = []\n",
    "    for i, seq,  in enumerate(variants):\n",
    "        results.append({\n",
    "            \"sequence\": seq,\n",
    "            \"Score_TRSp\": Score_TRSp[i],\n",
    "            \"Score_TRSpAvd\": Score_TRSpAvd[i],\n",
    "            \"Score_sum\": Score_TRSp[i] + Score_TRSpAvd[i]\n",
    "        })\n",
    "    return results\n",
    "\n",
    "def optimize_sequence(original_seq, frame_offset=0,CHANGES=6,rank_max=2,forbidden_positions=[],threshold=0.7,codon_usage=codon_usage_ecoli):\n",
    "    \"\"\"\n",
    "    Run beam search until one good variant is found.\n",
    "    \"\"\"\n",
    "    beam = [original_seq]\n",
    "    evaluated=evaluate_sequences(beam)\n",
    "    good = [v for v in evaluated if v[\"Score_TRSp\"] >= threshold and v[\"Score_TRSpAvd\"] >= threshold]\n",
    "    if good:\n",
    "        return {\n",
    "            \"Original_Sequence\": original_seq,\n",
    "            \"New_Variant\": evaluated[\"sequence\"],\n",
    "            \"Score_TRSp\": evaluated[\"Score_TRSp\"],\n",
    "            \"Score_TRSpAvd\": evaluated[\"Score_TRSpAvd\"]\n",
    "        }\n",
    "    codon_changes_done=0\n",
    "    while beam and codon_changes_done<=CHANGES:\n",
    "        codon_changes_done+=1\n",
    "        variants = []\n",
    "        for seq in beam:\n",
    "            variants.extend(propose_single_codon_changes(seq, frame_offset,rank_max,forbidden_positions,codon_usage))\n",
    "        variants=np.unique(variants)\n",
    "\n",
    "        evaluated = evaluate_sequences(variants)\n",
    "\n",
    "        good = [v for v in evaluated if v[\"Score_TRSp\"] >= threshold and v[\"Score_TRSpAvd\"] >= threshold]\n",
    "        if good:\n",
    "            best = max(good, key=lambda x: x[\"Score_sum\"])\n",
    "            return {\n",
    "                \"Original_Sequence\": original_seq,\n",
    "                \"New_Variant\": best[\"sequence\"],\n",
    "                \"Score_TRSp\": best[\"Score_TRSp\"],\n",
    "                \"Score_TRSpAvd\": best[\"Score_TRSpAvd\"]\n",
    "            }\n",
    "        elif codon_changes_done==CHANGES+1:\n",
    "            best = max(evaluated, key=lambda x: x[\"Score_sum\"])\n",
    "            return {\n",
    "                \"Original_Sequence\": original_seq,\n",
    "                \"New_Variant\": best[\"sequence\"],\n",
    "                \"Score_TRSp\": best[\"Score_TRSp\"],\n",
    "                \"Score_TRSpAvd\": best[\"Score_TRSpAvd\"]\n",
    "            }\n",
    "        \n",
    "\n",
    "        evaluated.sort(key=lambda x: x[\"Score_sum\"], reverse=True)\n",
    "        # Keep only Pareto-optimal variants\n",
    "        pareto_variants = pareto_front(evaluated)\n",
    "\n",
    "        # Optional: still restrict beam size if too many\n",
    "        pareto_variants.sort(key=lambda x: x[\"Score_sum\"], reverse=True)\n",
    "        beam = [v[\"sequence\"] for v in pareto_variants[:40]]\n",
    "\n",
    "\n",
    "    return {\n",
    "        \"Original_Sequence\": original_seq,\n",
    "        \"New_Variant\": original_seq,\n",
    "        \"All_Codon_Changes\": [],\n",
    "        \"Score_TRSp\": None,\n",
    "        \"Score_TRSpAvd\": None\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a173e204-a395-4013-929e-cb2b5475b70e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TCGGTTTCACCCCCAGTCTGCCGGGGTTCATACGCCCCGGCCCCTTAGGACCATCGGGGAAATCTTTATT\n"
     ]
    },
    {
     "ename": "TypeError",
     "evalue": "encode_tr_list() takes 1 positional argument but 2 were given",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[6], line 4\u001b[0m\n\u001b[1;32m      1\u001b[0m seq \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;241m.\u001b[39mjoin(np\u001b[38;5;241m.\u001b[39mrandom\u001b[38;5;241m.\u001b[39mchoice([\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mA\u001b[39m\u001b[38;5;124m'\u001b[39m,\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mT\u001b[39m\u001b[38;5;124m'\u001b[39m,\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mC\u001b[39m\u001b[38;5;124m'\u001b[39m,\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mG\u001b[39m\u001b[38;5;124m'\u001b[39m], size\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m70\u001b[39m))\n\u001b[1;32m      2\u001b[0m \u001b[38;5;28mprint\u001b[39m(seq)\n\u001b[0;32m----> 4\u001b[0m \u001b[43moptimize_sequence\u001b[49m\u001b[43m(\u001b[49m\u001b[43mseq\u001b[49m\u001b[43m)\u001b[49m\n",
      "Cell \u001b[0;32mIn[5], line 190\u001b[0m, in \u001b[0;36moptimize_sequence\u001b[0;34m(original_seq, frame_offset, CHANGES, rank_max, forbidden_positions, threshold, codon_usage)\u001b[0m\n\u001b[1;32m    186\u001b[0m \u001b[38;5;250m\u001b[39m\u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m    187\u001b[0m \u001b[38;5;124;03mRun beam search until one good variant is found.\u001b[39;00m\n\u001b[1;32m    188\u001b[0m \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m    189\u001b[0m beam \u001b[38;5;241m=\u001b[39m [original_seq]\n\u001b[0;32m--> 190\u001b[0m evaluated\u001b[38;5;241m=\u001b[39m\u001b[43mevaluate_sequences\u001b[49m\u001b[43m(\u001b[49m\u001b[43mbeam\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    191\u001b[0m good \u001b[38;5;241m=\u001b[39m [v \u001b[38;5;28;01mfor\u001b[39;00m v \u001b[38;5;129;01min\u001b[39;00m evaluated \u001b[38;5;28;01mif\u001b[39;00m v[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mScore_TRSp\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m>\u001b[39m\u001b[38;5;241m=\u001b[39m threshold \u001b[38;5;129;01mand\u001b[39;00m v[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mScore_TRSpAvd\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m>\u001b[39m\u001b[38;5;241m=\u001b[39m threshold]\n\u001b[1;32m    192\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m good:\n",
      "Cell \u001b[0;32mIn[5], line 166\u001b[0m, in \u001b[0;36mevaluate_sequences\u001b[0;34m(variants)\u001b[0m\n\u001b[1;32m    162\u001b[0m \u001b[38;5;250m\u001b[39m\u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m    163\u001b[0m \u001b[38;5;124;03mEncode sequences, run classifiers, return list of dicts with scores.\u001b[39;00m\n\u001b[1;32m    164\u001b[0m \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m    165\u001b[0m seqs \u001b[38;5;241m=\u001b[39m [v \u001b[38;5;28;01mfor\u001b[39;00m v \u001b[38;5;129;01min\u001b[39;00m variants]\n\u001b[0;32m--> 166\u001b[0m Scores\u001b[38;5;241m=\u001b[39m\u001b[43mencoding\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mencode_tr_list\u001b[49m\u001b[43m(\u001b[49m\u001b[43mseqs\u001b[49m\u001b[43m,\u001b[49m\u001b[38;5;241;43m2\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m    167\u001b[0m Score_TRSp \u001b[38;5;241m=\u001b[39m [a \u001b[38;5;28;01mfor\u001b[39;00m [a,b] \u001b[38;5;129;01min\u001b[39;00m Scores]\n\u001b[1;32m    168\u001b[0m Score_TRSpAvd \u001b[38;5;241m=\u001b[39m [b \u001b[38;5;28;01mfor\u001b[39;00m [a,b] \u001b[38;5;129;01min\u001b[39;00m Scores]\n",
      "\u001b[0;31mTypeError\u001b[0m: encode_tr_list() takes 1 positional argument but 2 were given"
     ]
    }
   ],
   "source": [
    "seq = ''.join(np.random.choice(['A','T','C','G'], size=70))\n",
    "print(seq)\n",
    "\n",
    "optimize_sequence(seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "31c6d692-099b-4704-8546-b1612bf9d159",
   "metadata": {},
   "outputs": [],
   "source": [
    "#| hide\n",
    "import nbdev; nbdev.nbdev_export()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "python3",
   "language": "python",
   "name": "python3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
