{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# evolution\n",
    "\n",
    "> We define all the usefull functions to evolve a bad TR to a good TR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| default_exp evolution"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def synonymous_mutation(sequence):\n",
    "    \"\"\"\n",
    "    Generates synonymous mutations for a given DNA sequence.\n",
    "\n",
    "    Parameters:\n",
    "        sequence (str): The input DNA sequence.\n",
    "\n",
    "    Returns:\n",
    "        list: A list of synonymous mutant sequences.\n",
    "    \"\"\"\n",
    "    # Define the genetic code (mapping codons to amino acids)\n",
    "    genetic_code = {\n",
    "        'TTT': 'F', 'TTC': 'F', 'TTA': 'L', 'TTG': 'L',\n",
    "        'TCT': 'S', 'TCC': 'S', 'TCA': 'S', 'TCG': 'S',\n",
    "        'TAT': 'Y', 'TAC': 'Y', 'TAA': '*', 'TAG': '*',\n",
    "        'TGT': 'C', 'TGC': 'C', 'TGA': '*', 'TGG': 'W',\n",
    "        'CTT': 'L', 'CTC': 'L', 'CTA': 'L', 'CTG': 'L',\n",
    "        'CCT': 'P', 'CCC': 'P', 'CCA': 'P', 'CCG': 'P',\n",
    "        'CAT': 'H', 'CAC': 'H', 'CAA': 'Q', 'CAG': 'Q',\n",
    "        'CGT': 'R', 'CGC': 'R', 'CGA': 'R', 'CGG': 'R',\n",
    "        'ATT': 'I', 'ATC': 'I', 'ATA': 'I', 'ATG': 'M',\n",
    "        'ACT': 'T', 'ACC': 'T', 'ACA': 'T', 'ACG': 'T',\n",
    "        'AAT': 'N', 'AAC': 'N', 'AAA': 'K', 'AAG': 'K',\n",
    "        'AGT': 'S', 'AGC': 'S', 'AGA': 'R', 'AGG': 'R',\n",
    "        'GTT': 'V', 'GTC': 'V', 'GTA': 'V', 'GTG': 'V',\n",
    "        'GCT': 'A', 'GCC': 'A', 'GCA': 'A', 'GCG': 'A',\n",
    "        'GAT': 'D', 'GAC': 'D', 'GAA': 'E', 'GAG': 'E',\n",
    "        'GGT': 'G', 'GGC': 'G', 'GGA': 'G', 'GGG': 'G'\n",
    "    }\n",
    "\n",
    "    mutant_sequences = [sequence]\n",
    "\n",
    "    # Iterate over the input sequence in triplets (codons)\n",
    "    for i in range(0, len(sequence), 3):\n",
    "        codon = sequence[i:i+3]\n",
    "        # print(codon)\n",
    "\n",
    "        # Iterate over all possible mutations of the codon\n",
    "        for j in range(3):\n",
    "            for base in 'ACGT':\n",
    "                if base != codon[j]:\n",
    "                    mutated_codon = codon[:j] + base + codon[j+1:]\n",
    "                    # Check if mutation results in the same amino acid\n",
    "                    if genetic_code[codon] == genetic_code.get(mutated_codon, ''):\n",
    "                        mutant_sequence = sequence[:i] + mutated_codon + sequence[i+3:]\n",
    "                        mutant_sequences.append(mutant_sequence)\n",
    "\n",
    "    return mutant_sequences\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def translate_dna_to_protein(dna_sequence):\n",
    "    codon_table = {\n",
    "        'ATA': 'I', 'ATC': 'I', 'ATT': 'I', 'ATG': 'M',\n",
    "        'ACA': 'T', 'ACC': 'T', 'ACG': 'T', 'ACT': 'T',\n",
    "        'AAC': 'N', 'AAT': 'N', 'AAA': 'K', 'AAG': 'K',\n",
    "        'AGC': 'S', 'AGT': 'S', 'AGA': 'R', 'AGG': 'R',\n",
    "        'CTA': 'L', 'CTC': 'L', 'CTG': 'L', 'CTT': 'L',\n",
    "        'CCA': 'P', 'CCC': 'P', 'CCG': 'P', 'CCT': 'P',\n",
    "        'CAC': 'H', 'CAT': 'H', 'CAA': 'Q', 'CAG': 'Q',\n",
    "        'CGA': 'R', 'CGC': 'R', 'CGG': 'R', 'CGT': 'R',\n",
    "        'GTA': 'V', 'GTC': 'V', 'GTG': 'V', 'GTT': 'V',\n",
    "        'GCA': 'A', 'GCC': 'A', 'GCG': 'A', 'GCT': 'A',\n",
    "        'GAC': 'D', 'GAT': 'D', 'GAA': 'E', 'GAG': 'E',\n",
    "        'GGA': 'G', 'GGC': 'G', 'GGG': 'G', 'GGT': 'G',\n",
    "        'TCA': 'S', 'TCC': 'S', 'TCG': 'S', 'TCT': 'S',\n",
    "        'TTC': 'F', 'TTT': 'F', 'TTA': 'L', 'TTG': 'L',\n",
    "        'TAC': 'Y', 'TAT': 'Y', 'TAA': '*', 'TAG': '*',\n",
    "        'TGC': 'C', 'TGT': 'C', 'TGA': '*', 'TGG': 'W',\n",
    "    }\n",
    "    \n",
    "    protein_sequence = ''\n",
    "    \n",
    "    for i in range(0, len(dna_sequence), 3):\n",
    "        codon = dna_sequence[i:i+3]\n",
    "        if codon in codon_table:\n",
    "            protein_sequence += codon_table[codon]\n",
    "        else:\n",
    "            protein_sequence += 'X'  # Placeholder for unknown codons\n",
    "    \n",
    "    return protein_sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def synonymous_mutation_list(seq_list):\n",
    "    syn_list=[]\n",
    "    for seq in seq_list:\n",
    "        syn_list.extend(synonymous_mutation(seq))\n",
    "    return syn_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def framing_TR(gene,list_TR):\n",
    "   \n",
    "    start_frame_list=[]\n",
    "    end_frame_list=[]\n",
    "    base_start_list=[]\n",
    "    base_end_list=[]\n",
    "    extended_TR_list=[]\n",
    "    for k in list_TR:\n",
    "        TR_size=len(k)\n",
    "        start_position=gene.index(k)+1\n",
    "        end_position=start_position+TR_size-1\n",
    "        # print(start_position,end_position)\n",
    "        if start_position%3==1:\n",
    "            start_frame=0\n",
    "        elif start_position%3==2:\n",
    "            start_frame=1\n",
    "        elif start_position%3==0:\n",
    "            start_frame=2\n",
    "        if end_position%3==1:\n",
    "            end_frame=2\n",
    "        elif end_position%3==2:\n",
    "            end_frame=1\n",
    "        elif end_position%3==0:\n",
    "            end_frame=0\n",
    "\n",
    "        base_start=gene[start_position-1-start_frame:start_position-1]\n",
    "        base_end=gene[end_position:end_position+end_frame]\n",
    "        extended_TR=base_start+k+base_end\n",
    "        \n",
    "        start_frame_list.append(start_frame)\n",
    "        end_frame_list.append(end_frame)\n",
    "        base_start_list.append(base_start)\n",
    "        base_end_list.append(base_end)\n",
    "        extended_TR_list.append(extended_TR)\n",
    "        \n",
    "    return (start_frame_list,end_frame_list,base_start_list,base_end_list,extended_TR_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def evolution_synonymous_TR_forward_GENE_forward(TR,gene,iteration_max,score_goal,model):\n",
    "    len_DNA=len(TR)\n",
    "    # print(len_DNA)\n",
    "    X_0=double_encode([TR])\n",
    "    score_0 = model.predict_proba(X_0)[:,1]    \n",
    "    print('Round 0','Score =',np.round(score_0[0],decimals=2))\n",
    "    history_mutants=[[TR]]\n",
    "\n",
    "    n_base_to_add_start=framing_TR(gene,[TR])[0][0]\n",
    "    n_base_to_add_end=framing_TR(gene,[TR])[1][0]    \n",
    "    base_to_add_start=framing_TR(gene,[TR])[2][0]\n",
    "    base_to_add_end=framing_TR(gene,[TR])[3][0]\n",
    "    extended_gene=framing_TR(gene,[TR])[4][0]\n",
    "    \n",
    "    # print('base_to_add_start',base_to_add_start,'base_to_add_end',base_to_add_end)\n",
    "\n",
    "    DNA_sequence=[TR]\n",
    "\n",
    "    score_best=score_0\n",
    "    iteration=0\n",
    "\n",
    "    # for k in range (iteration_max):\n",
    "    while (score_best<score_goal) and (iteration<iteration_max):\n",
    "        #1 Putting the DNA sequence in frame\n",
    "        in_frame_DNA_sequence=[base_to_add_start+seq+base_to_add_end for seq in DNA_sequence]\n",
    "        print(len(in_frame_DNA_sequence[0]))\n",
    "\n",
    "\n",
    "        #3 Generation all the synonymous mutants of the in frame DNA sequence\n",
    "        mutant_list_in_frame=synonymous_mutation_list(in_frame_DNA_sequence)\n",
    "\n",
    "        #5 Restoring the outframe DNA sequence \n",
    "        out_frame_mutant_list_forward=[seq[n_base_to_add_start:len_DNA+n_base_to_add_start] for seq in mutant_list_in_frame]\n",
    "\n",
    "        #6 Scoring the different sequences\n",
    "        X = double_encode(out_frame_mutant_list_forward)\n",
    "        score = model.predict_proba(X)[:,1]\n",
    "\n",
    "        #7 Taking only the best sequences\n",
    "        score_best=np.mean(np.sort(score)[-3:])  \n",
    "        print('Round =',iteration+1,'Best Score=',np.round(score_best,decimals=2))\n",
    "    \n",
    "        df=pd.DataFrame({'Sequence':out_frame_mutant_list_forward,'Score':score })\n",
    "        df_best=df.sort_values(by='Score',ascending=False).iloc[:5]\n",
    "        \n",
    "        DNA_sequence=list(df_best['Sequence'])\n",
    "        history_mutants.append(DNA_sequence)\n",
    "\n",
    "        # colored_mismatch=[]\n",
    "        # print(DNA_sequence)\n",
    "        # for j in DNA_sequence:\n",
    "        #     print(show_mismatches(TR,j)[0])\n",
    "        #     print(show_mismatches(TR,j)[1])\n",
    "        #     print('------')\n",
    "        iteration+=1\n",
    "\n",
    "    return df_best\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "python3",
   "language": "python",
   "name": "python3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
