<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.8.27">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="description" content="We define all the usefull functions to predict the potential VR from a given TR.">

<title>lstm – dgrec</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<script src="../site_libs/quarto-html/quarto.js" type="module"></script>
<script src="../site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="../site_libs/quarto-html/axe/axe-check.js" type="module"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-ed96de9b727972fe78a7b5d16c58bf87.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap-000f0825b9c55ed21d14970d810c14a9.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="../styles.css">
<meta property="og:title" content="lstm – dgrec">
<meta property="og:description" content="We define all the usefull functions to predict the potential VR from a given TR.">
<meta property="og:image" content="https://dbikard.github.io/dgrec/API/09_lstm_files/figure-html/cell-6-output-4.png">
<meta property="og:site_name" content="dgrec">
<meta property="og:image:height" content="240">
<meta property="og:image:width" content="989">
<meta name="twitter:title" content="lstm – dgrec">
<meta name="twitter:description" content="We define all the usefull functions to predict the potential VR from a given TR.">
<meta name="twitter:image" content="https://dbikard.github.io/dgrec/API/09_lstm_files/figure-html/cell-6-output-4.png">
<meta name="twitter:image-height" content="240">
<meta name="twitter:image-width" content="989">
<meta name="twitter:card" content="summary_large_image">
</head>

<body class="nav-sidebar floating nav-fixed quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">dgrec</span>
    </a>
  </div>
        <div class="quarto-navbar-tools tools-end">
</div>
          <div id="quarto-search" class="" title="Search"></div>
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../API/genotypes.html">API</a></li><li class="breadcrumb-item"><a href="../API/lstm.html">lstm</a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">dgrec</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true">
 <span class="menu-text">API</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../API/genotypes.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">genotypes</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../API/genotypes_paired.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">genotypes_paired</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../API/plotting.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">plotting</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../API/analysis.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">analysis</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../API/encoding.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">encoding</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../API/predictions.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">predictions</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../API/utils.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">utils</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../API/cli.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Command line interface</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../API/lstm.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">lstm</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true">
 <span class="menu-text">DGRec data analysis</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../DGRec data analysis/mutation_distribution.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Mutations distribution</span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#generate_sequences_onetr" id="toc-generate_sequences_onetr" class="nav-link active" data-scroll-target="#generate_sequences_onetr">generate_sequences_oneTR</a></li>
  <li><a href="#generate_sequences" id="toc-generate_sequences" class="nav-link" data-scroll-target="#generate_sequences">generate_sequences</a></li>
  <li><a href="#evaluatetr_to_prot" id="toc-evaluatetr_to_prot" class="nav-link" data-scroll-target="#evaluatetr_to_prot">EvaluateTR_to_prot</a></li>
  <li><a href="#optimize_sequence_display_proteins" id="toc-optimize_sequence_display_proteins" class="nav-link" data-scroll-target="#optimize_sequence_display_proteins">optimize_sequence_display_proteins</a></li>
  <li><a href="#compute_likelihood" id="toc-compute_likelihood" class="nav-link" data-scroll-target="#compute_likelihood">compute_likelihood</a></li>
  <li><a href="#compute_likelihood_list" id="toc-compute_likelihood_list" class="nav-link" data-scroll-target="#compute_likelihood_list">compute_likelihood_list</a></li>
  <li><a href="#compute_likelihood_matrix" id="toc-compute_likelihood_matrix" class="nav-link" data-scroll-target="#compute_likelihood_matrix">compute_likelihood_matrix</a></li>
  </ul>
<div class="toc-actions"><ul><li><a href="https://github.com/dbikard/dgrec/issues/new" class="toc-action"><i class="bi bi-github"></i>Report an issue</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../API/genotypes.html">API</a></li><li class="breadcrumb-item"><a href="../API/lstm.html">lstm</a></li></ol></nav>
<div class="quarto-title">
<h1 class="title">lstm</h1>
</div>

<div>
  <div class="description">
    We define all the usefull functions to predict the potential VR from a given TR.
  </div>
</div>


<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<!-- WARNING: THIS FILE WAS AUTOGENERATED! DO NOT EDIT! -->
<hr>
<p><a href="https://github.com/dbikard/dgrec/blob/main/dgrec/lstm.py#L424" target="_blank" style="float:right; font-size:smaller">source</a></p>
<section id="generate_sequences_onetr" class="level3">
<h3 class="anchored" data-anchor-id="generate_sequences_onetr">generate_sequences_oneTR</h3>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> generate_sequences_oneTR(</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>    TR:<span class="bu">str</span>, <span class="co"># TR sequence</span></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>    n:<span class="bu">int</span><span class="op">=</span><span class="dv">1000</span>, <span class="co"># NUmber of VR to generate</span></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>)<span class="op">-&gt;</span><span class="bu">list</span>: <span class="co"># List of `n` generated VR sequences.</span></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p><em>Generate multiple VR sequences from a single TR sequence.</em></p>
<hr>
<p><a href="https://github.com/dbikard/dgrec/blob/main/dgrec/lstm.py#L369" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="generate_sequences" class="level3">
<h3 class="anchored" data-anchor-id="generate_sequences">generate_sequences</h3>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb2"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> generate_sequences(</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>    X_seq:<span class="bu">list</span>, <span class="co"># list of TR sequences</span></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>)<span class="op">-&gt;</span><span class="bu">list</span>: <span class="co"># List of generated VR sequences, in the same order as `X_seq`.</span></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p><em>Generate VR sequences from a list of TR sequences.</em></p>
<p>Each TR sequence produces exactly one VR sequence.</p>
<div id="d6906e1a-489b-43f0-80b5-58af42453615" class="cell" data-execution_count="6">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb3"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>generate_sequences_oneTR(<span class="st">'CGTAAACCGGACCTAGTTTAGTTCTTAGACCAAGGTACATATCCCCGTAACATAAGACGCGACTGGGCCC'</span>,n<span class="op">=</span><span class="dv">10</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>2026-01-29 14:31:02.870801: I external/local_xla/xla/stream_executor/cuda/cuda_dnn.cc:531] Loaded cuDNN version 8907</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>1/1 ━━━━━━━━━━━━━━━━━━━━ 4s 4s/step</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Generating sequence: 100%|██████████████████████| 70/70 [00:18&lt;00:00,  3.73it/s]</code></pre>
</div>
<div class="cell-output cell-output-display" data-execution_count="6">
<pre><code>['CGTAAACCGGACCTAGTTTAGTTCTTGGACCAAGGTACATATCCCCGTCACATAAGACGCGACTGGGCCC',
 'CGTAGACCGGTCCTAGTTTGGTTCTTAGACCAAGGTACATATCCCCGTAACATAAGACGCGACCGGGCCC',
 'CGTAAACCGGCCCTAGTTTAGTTCTTAGACCAAGGTACATAGACCCGTGCCATAAGACGCGACTGGGCCC',
 'CGTAAACCGGGCCTAGTTTAGGTCTTAGTCCAAGGTACATGTCCCCGTTACATAAGGCGCGACTGGGCCC',
 'CGTAAACCGGACCTAGTTTAGTTCTTAGCCCAGGGTACGTATACCCGTACCATAAGACGCGACTGGGCCC',
 'CGTAAACCGGACCTAGTTTAGTTCTTAGACCAGGGTGCATGCCCCCGTTACATAGGACGCGACTGGGCCC',
 'CGTTTACCGGACCTGGTTTAGTTCTTAGGCCAAGGTACTTATCCCCGTGGCATGGGACGCGACTGGGCCC',
 'CGTAAACCGGACCTAGTTTAGTTCTTAGACCAAGGTACATATCCACGTAACATAAGGCGCGACTGGGCCC',
 'CGTAACCCGGACCTAGTTTAGTTCTTGGACCACGGGACATATCCCCGTAACATAAGCCGCGACTGGGCCC',
 'CGTACCCCGGACCTAGTTTAGTTCTTAGACCAAGGTACATATCCCCGTAACATAAGACGCGACTGGGCCC']</code></pre>
</div>
</div>
<hr>
<p><a href="https://github.com/dbikard/dgrec/blob/main/dgrec/lstm.py#L446" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="evaluatetr_to_prot" class="level3">
<h3 class="anchored" data-anchor-id="evaluatetr_to_prot">EvaluateTR_to_prot</h3>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb8"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> EvaluateTR_to_prot(</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>    TR:<span class="bu">str</span>, <span class="co"># The TR sequence</span></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>    NDGR:<span class="bu">int</span><span class="op">=</span><span class="dv">100</span>, <span class="co"># Number of VR to generate for the protein logo</span></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>    offset:<span class="bu">int</span><span class="op">=</span><span class="dv">0</span>, <span class="co"># The offset for protein translation</span></span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>)<span class="op">-&gt;</span>Counter: <span class="co"># Counts of translated protein sequences.</span></span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p><em>Evaluate protein diversity accessible from a TR sequence via DGR.</em></p>
<p>Generates VR sequences from a single TR, translates them into proteins, and displays a protein sequence logo based on amino-acid frequencies.</p>
<div id="b9724c59-9a7c-4c35-96dc-8ecd744dffc6" class="cell" data-scrolled="true" data-execution_count="8">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb9"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>EvaluateTR_to_prot(<span class="st">'CGTAAACCGGACCTAGTTTAGTTCTTAGACCAAGGTACATATCCCCGTAACATAAGACGCGACTGGGCCC'</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>4/4 ━━━━━━━━━━━━━━━━━━━━ 3s 32ms/step</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Generating sequence: 100%|██████████████████████| 70/70 [00:19&lt;00:00,  3.63it/s]
/home/regnier/tf_env/lib/python3.10/site-packages/logomaker/src/error_handling.py:58: UserWarning:  Warning: Character '*' is not in color_dict. Using black.
  warnings.warn(str(Error))</code></pre>
</div>
<div class="cell-output cell-output-display">
<pre><code>&lt;Figure size 1200x400 with 0 Axes&gt;</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="09_lstm_files/figure-html/cell-6-output-4.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="8">
<pre><code>Counter({'RKPDLV*FLDQGTYPRNIRRDWA': 7,
         'RKPDLV*FLDQGTYPRYIRRDWA': 5,
         'RKPDLV*FLDQGPYPRNIRRDWA': 4,
         'RKPDLV*FLAQGTYPRNIRRDWA': 3,
         'RKPDLVSFLDQGTYPRNIRRDWA': 3,
         'RKPDLV*FLDQGTYPRHIRRDWA': 3,
         'RKPGLV*FLDQGTYPRNIRRDWA': 2,
         'RKPDLV*FLDQGTYPRSIRRDWA': 2,
         'RKPDLV*FLDQGTSPRNIRRDWA': 2,
         'RIPDLV*FLDQGAYPRSVRRDWA': 1,
         'RKPDLV*FFDHGTYPRNIRRDWA': 1,
         'RKPALV*FLDQGTYPRNIRRDWA': 1,
         'RKPDLV*FLAQGPYPRSLRRDWA': 1,
         'RKPDLVSFLDQCTYPRNIRRDWA': 1,
         'RVPVLV*FLDQGTYPRNIRRDWA': 1,
         'RKPDLV*FFDQGTYPRNIRRDWA': 1,
         'RKPDLV*FLALGTYPRDIRRDWA': 1,
         'RKPALV*FLAQGTCPRYMRRDWA': 1,
         'RKPDLV*FLDQGTYPRYISRDWA': 1,
         'RIPDLV*FLAQGACPRNIGRAWA': 1,
         'RIPALVWFFGQGSYPRNIGRAWA': 1,
         'RKPVLVWFLDQGTSPRYIRRDWA': 1,
         'RKPDLV*FLAQGAYPRIIGRDWA': 1,
         'RKPDLV*FLGQGTYPRNMGRDWA': 1,
         'RKPGLVEFLAQGAYPRYIRRDWA': 1,
         'RKPDLV*FLAQGACPRNIRRDWA': 1,
         'RLPALV*FLDQGTYPRYIRRDWA': 1,
         'RSPDLV*FLDRGAYPRNLRRDWA': 1,
         'RKPDLV*FLDQGTFPRYIRRDWA': 1,
         'RKPDLV*FLDQGTCPRYISRDWA': 1,
         'RKPVLV*FLDQGTYPRNIRRDWA': 1,
         'RKPVLVSFLDQGTYPRNIRRDWA': 1,
         'RKPDLV*FLDQGTGPRYIRRDWA': 1,
         'RNPDLV*FLDQGTCPRNIRRDWA': 1,
         'RKPDLVWVLDQGTCPRNIRRDWA': 1,
         'RKPDLV*FLAHGTYPRNIRRDWA': 1,
         'RIPDLV*FLDQGTYPRNIRRDWA': 1,
         'RKPDLV*FLDQGTYPRNMGRDWA': 1,
         'RKPDLV*FLDQGTCHRSIRRDWA': 1,
         'RKPDLV*FLDQGTFPRNIRRDWA': 1,
         'RKPDLV*FLDQGAYPRFIRRDWA': 1,
         'RKPDLV*FLVQGAYPRCMRRDWA': 1,
         'RKPDLV*FLDQGAYPRGLRRDWA': 1,
         'RKPDLVSFLDQGTYPRYIRRDWA': 1,
         'RKPDLV*FLDQGAYPRGIRRAWA': 1,
         'RKPDLA*FLDRGPWPRDIRRDWA': 1,
         'RKPDLV*FFDQGSYPRNIRRDWA': 1,
         'RKPDLVWFLAQGAYPRYI*RDWA': 1,
         'RKPDLV*FLDQGTYPRNIGRDWA': 1,
         'RNPDLV*FLDQVTWPRSVRRDWA': 1,
         'RIPDLV*FLDQGTSPRDIRRDWA': 1,
         'RKPDLV*FLVQGACPRSIRRDWA': 1,
         'RLPDLV*FLDQGACPRSVRRDWA': 1,
         'RFPVLV*FFAQGACPRYIRRDWA': 1,
         'RKPDLV*VFAQGAYPRYIRRDWA': 1,
         'RKPVLV*FLDQGTSPRTMRRDWA': 1,
         'RKPALV*FLAQGPYPRNIRRDWA': 1,
         'RKPDLV*FLDPGTYPRYIRRDWA': 1,
         'RKPDLV*FLDHGTYPRNIRRDWA': 1,
         'RKPDLV*FLAQGTYPRSIRRDWA': 1,
         'RKPDLV*FLAQGTCPRSIRRDWA': 1,
         'RKPDLV*FLDQGTYPRHLRRDWA': 1,
         'RKPALV*FLDQGSWPRYIRRDWA': 1,
         'RKPDLV*FLAQGTYPRFIRRDWA': 1,
         'RKPALV*FLDRGTYPRGIRRDWA': 1,
         'RKPDLVWFLVRGAYPRDVRRDWA': 1,
         'RKPDLV*FLDQGTFPRYISRDSA': 1,
         'RKPDLV*FLAPGTYPRNIRRAWA': 1,
         'RLPVLV*FLARGTYPRGIRRDWA': 1,
         'RKPALVWFLDQGPYPRLIRRDWA': 1,
         'RKPALV*FLAQGTYPRNIRRDWA': 1,
         'RKPVLVWFLDQGTYPRNIRRDWA': 1,
         'RKPDLV*FLALGTYPRNIRRDWA': 1,
         'RKPGLV*FLDQGTWPRSIRRDWA': 1,
         'RQPDLVSFLDQGTYPRNIRRDWA': 1,
         'RKPDLV*FLDRGPCPRGIRRDWA': 1,
         'RSPDLV*FLDQGTYPRNIRRDWA': 1,
         'RKPDLV*FLAQGSYPRNIRRDWA': 1})</code></pre>
</div>
</div>
<hr>
<p><a href="https://github.com/dbikard/dgrec/blob/main/dgrec/lstm.py#L511" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="optimize_sequence_display_proteins" class="level3">
<h3 class="anchored" data-anchor-id="optimize_sequence_display_proteins">optimize_sequence_display_proteins</h3>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb14"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> optimize_sequence_display_proteins(</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>    original_seq:<span class="bu">str</span>, <span class="co"># Original DNA sequence to optimize.</span></span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>    frame_offset:<span class="bu">int</span><span class="op">=</span><span class="dv">0</span>, <span class="co"># Reading-frame offset (0, 1, or 2) used when grouping codons.</span></span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>    CHANGES:<span class="bu">int</span><span class="op">=</span><span class="dv">6</span>, <span class="co"># Maximum number of codon substitutions allowed.</span></span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>    freq_min:<span class="bu">float</span><span class="op">=</span><span class="fl">0.2</span>, <span class="co"># Lowest usage frequency acceptable.</span></span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a>    N:<span class="bu">int</span><span class="op">=</span><span class="dv">1</span>, <span class="co"># Number of putative TR to output.</span></span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a>    forbidden_positions:<span class="bu">list</span><span class="op">=</span>[], <span class="co"># Nucleotide positions that must not be modified.</span></span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a>    threshold:<span class="bu">float</span><span class="op">=</span><span class="fl">0.7</span>, <span class="co"># Minimum required value for both `Score_TRSp` and `Score_TRSpAvd` to</span></span>
<span id="cb14-10"><a href="#cb14-10" aria-hidden="true" tabindex="-1"></a>accept a sequence <span class="im">as</span> optimal.</span>
<span id="cb14-11"><a href="#cb14-11" aria-hidden="true" tabindex="-1"></a>    codon_usage:<span class="bu">dict</span><span class="op">=</span>{<span class="st">'F'</span>: {<span class="st">'TTT'</span>: <span class="fl">0.57</span>, <span class="st">'TTC'</span>: <span class="fl">0.43</span>}, <span class="st">'L'</span>: {<span class="st">'TTA'</span>: <span class="fl">0.15</span>, <span class="st">'TTG'</span>: <span class="fl">0.12</span>, <span class="st">'CTT'</span>: <span class="fl">0.12</span>, <span class="st">'CTC'</span>: <span class="fl">0.1</span>, <span class="st">'CTA'</span>: <span class="fl">0.05</span>, <span class="st">'CTG'</span>: <span class="fl">0.46</span>}, <span class="st">'S'</span>: {<span class="st">'TCT'</span>: <span class="fl">0.11</span>, <span class="st">'TCC'</span>: <span class="fl">0.11</span>, <span class="st">'TCA'</span>: <span class="fl">0.15</span>, <span class="st">'TCG'</span>: <span class="fl">0.16</span>, <span class="st">'AGT'</span>: <span class="fl">0.14</span>, <span class="st">'AGC'</span>: <span class="fl">0.33</span>}, <span class="st">'Y'</span>: {<span class="st">'TAT'</span>: <span class="fl">0.53</span>, <span class="st">'TAC'</span>: <span class="fl">0.47</span>}, <span class="st">'*'</span>: {<span class="st">'TAA'</span>: <span class="fl">0.64</span>, <span class="st">'TAG'</span>: <span class="fl">0.0</span>, <span class="st">'TGA'</span>: <span class="fl">0.36</span>}, <span class="st">'C'</span>: {<span class="st">'TGT'</span>: <span class="fl">0.42</span>, <span class="st">'TGC'</span>: <span class="fl">0.58</span>}, <span class="st">'W'</span>: {<span class="st">'TGG'</span>: <span class="fl">1.0</span>}, <span class="st">'P'</span>: {<span class="st">'CCT'</span>: <span class="fl">0.17</span>, <span class="st">'CCC'</span>: <span class="fl">0.13</span>, <span class="st">'CCA'</span>: <span class="fl">0.14</span>, <span class="st">'CCG'</span>: <span class="fl">0.55</span>}, <span class="st">'H'</span>: {<span class="st">'CAT'</span>: <span class="fl">0.55</span>, <span class="st">'CAC'</span>: <span class="fl">0.45</span>}, <span class="st">'Q'</span>: {<span class="st">'CAA'</span>: <span class="fl">0.3</span>, <span class="st">'CAG'</span>: <span class="fl">0.7</span>}, <span class="st">'R'</span>: {<span class="st">'CGT'</span>: <span class="fl">0.36</span>, <span class="st">'CGC'</span>: <span class="fl">0.44</span>, <span class="st">'CGA'</span>: <span class="fl">0.07</span>, <span class="st">'CGG'</span>: <span class="fl">0.07</span>, <span class="st">'AGA'</span>: <span class="fl">0.07</span>, <span class="st">'AGG'</span>: <span class="fl">0.0</span>}, <span class="st">'I'</span>: {<span class="st">'ATT'</span>: <span class="fl">0.58</span>, <span class="st">'ATC'</span>: <span class="fl">0.35</span>, <span class="st">'ATA'</span>: <span class="fl">0.07</span>}, <span class="st">'M'</span>: {<span class="st">'ATG'</span>: <span class="fl">1.0</span>}, <span class="st">'T'</span>: {<span class="st">'ACT'</span>: <span class="fl">0.16</span>, <span class="st">'ACC'</span>: <span class="fl">0.47</span>, <span class="st">'ACA'</span>: <span class="fl">0.13</span>, <span class="st">'ACG'</span>: <span class="fl">0.24</span>}, <span class="st">'N'</span>: {<span class="st">'AAT'</span>: <span class="fl">0.47</span>, <span class="st">'AAC'</span>: <span class="fl">0.53</span>}, <span class="st">'K'</span>: {<span class="st">'AAA'</span>: <span class="fl">0.73</span>, <span class="st">'AAG'</span>: <span class="fl">0.27</span>}, <span class="st">'V'</span>: {<span class="st">'GTT'</span>: <span class="fl">0.25</span>, <span class="st">'GTC'</span>: <span class="fl">0.18</span>, <span class="st">'GTA'</span>: <span class="fl">0.17</span>, <span class="st">'GTG'</span>: <span class="fl">0.4</span>}, <span class="st">'A'</span>: {<span class="st">'GCT'</span>: <span class="fl">0.11</span>, <span class="st">'GCC'</span>: <span class="fl">0.31</span>, <span class="st">'GCA'</span>: <span class="fl">0.2</span>, <span class="st">'GCG'</span>: <span class="fl">0.38</span>}, <span class="st">'D'</span>: {<span class="st">'GAT'</span>: <span class="fl">0.65</span>, <span class="st">'GAC'</span>: <span class="fl">0.35</span>}, <span class="st">'E'</span>: {<span class="st">'GAA'</span>: <span class="fl">0.7</span>, <span class="st">'GAG'</span>: <span class="fl">0.3</span>}, <span class="st">'G'</span>: {<span class="st">'GGT'</span>: <span class="fl">0.29</span>, <span class="st">'GGC'</span>: <span class="fl">0.46</span>, <span class="st">'GGA'</span>: <span class="fl">0.13</span>, <span class="st">'GGG'</span>: <span class="fl">0.12</span>}}, <span class="co"># Codon usage table of E. Coli mapping amino acids to codons and frequencies.</span></span>
<span id="cb14-12"><a href="#cb14-12" aria-hidden="true" tabindex="-1"></a>    NDGR:<span class="bu">int</span><span class="op">=</span><span class="dv">100</span>, <span class="co"># Number of sequences to generate via the LSTM for sequence logo estimation. </span></span>
<span id="cb14-13"><a href="#cb14-13" aria-hidden="true" tabindex="-1"></a>):</span>
<span id="cb14-14"><a href="#cb14-14" aria-hidden="true" tabindex="-1"></a></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p><em>Optimize a DNA sequence via synonymous codon substitutions and shows the sequence logo for each of the optimal sequences.</em></p>
<p>This function performs a beam-search–based optimization of a nucleotide sequence by iteratively proposing single-codon synonymous changes and evaluating them with the two scoring functions. The search stops early if a variant meets the specified score thresholds, otherwise the best Pareto- optimal solution is returned.</p>
<div id="bf4d5240-9b07-43cf-8581-75019a28e806" class="cell" data-execution_count="10">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb15"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>seq <span class="op">=</span> <span class="st">''</span>.join(np.random.choice([<span class="st">'A'</span>,<span class="st">'T'</span>,<span class="st">'C'</span>,<span class="st">'G'</span>], size<span class="op">=</span><span class="dv">70</span>))</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(seq)</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>optimize_sequence_display_proteins(seq,N<span class="op">=</span><span class="dv">5</span>,CHANGES<span class="op">=</span><span class="dv">6</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>TAGATGCCGATTCATTCTAGCTTCGACAACAGTCAGCGAAAACTCTACCCACAGTAAATATTTCAGCAAG</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="09_lstm_files/figure-html/cell-8-output-2.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="09_lstm_files/figure-html/cell-8-output-3.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="09_lstm_files/figure-html/cell-8-output-4.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="09_lstm_files/figure-html/cell-8-output-5.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="09_lstm_files/figure-html/cell-8-output-6.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="09_lstm_files/figure-html/cell-8-output-7.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="09_lstm_files/figure-html/cell-8-output-8.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="09_lstm_files/figure-html/cell-8-output-9.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>4/4 ━━━━━━━━━━━━━━━━━━━━ 0s 30ms/step</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Generating sequence: 100%|██████████████████████| 70/70 [00:10&lt;00:00,  6.77it/s]
/home/regnier/tf_env/lib/python3.10/site-packages/logomaker/src/error_handling.py:58: UserWarning:  Warning: Character '*' is not in color_dict. Using black.
  warnings.warn(str(Error))</code></pre>
</div>
<div class="cell-output cell-output-display">
<pre><code>&lt;Figure size 1200x400 with 0 Axes&gt;</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="09_lstm_files/figure-html/cell-8-output-13.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>4/4 ━━━━━━━━━━━━━━━━━━━━ 0s 23ms/step</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Generating sequence: 100%|██████████████████████| 70/70 [00:16&lt;00:00,  4.25it/s]
/home/regnier/tf_env/lib/python3.10/site-packages/logomaker/src/error_handling.py:58: UserWarning:  Warning: Character '*' is not in color_dict. Using black.
  warnings.warn(str(Error))</code></pre>
</div>
<div class="cell-output cell-output-display">
<pre><code>&lt;Figure size 1200x400 with 0 Axes&gt;</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="09_lstm_files/figure-html/cell-8-output-17.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>4/4 ━━━━━━━━━━━━━━━━━━━━ 0s 15ms/step</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Generating sequence: 100%|██████████████████████| 70/70 [00:18&lt;00:00,  3.81it/s]
/home/regnier/tf_env/lib/python3.10/site-packages/logomaker/src/error_handling.py:58: UserWarning:  Warning: Character '*' is not in color_dict. Using black.
  warnings.warn(str(Error))</code></pre>
</div>
<div class="cell-output cell-output-display">
<pre><code>&lt;Figure size 1200x400 with 0 Axes&gt;</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="09_lstm_files/figure-html/cell-8-output-21.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>4/4 ━━━━━━━━━━━━━━━━━━━━ 0s 35ms/step</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Generating sequence: 100%|██████████████████████| 70/70 [00:10&lt;00:00,  6.84it/s]
/home/regnier/tf_env/lib/python3.10/site-packages/logomaker/src/error_handling.py:58: UserWarning:  Warning: Character '*' is not in color_dict. Using black.
  warnings.warn(str(Error))</code></pre>
</div>
<div class="cell-output cell-output-display">
<pre><code>&lt;Figure size 1200x400 with 0 Axes&gt;</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="09_lstm_files/figure-html/cell-8-output-25.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>4/4 ━━━━━━━━━━━━━━━━━━━━ 0s 31ms/step</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Generating sequence: 100%|██████████████████████| 70/70 [00:25&lt;00:00,  2.80it/s]
/home/regnier/tf_env/lib/python3.10/site-packages/logomaker/src/error_handling.py:58: UserWarning:  Warning: Character '*' is not in color_dict. Using black.
  warnings.warn(str(Error))</code></pre>
</div>
<div class="cell-output cell-output-display">
<pre><code>&lt;Figure size 1200x400 with 0 Axes&gt;</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="09_lstm_files/figure-html/cell-8-output-29.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="10">
<pre><code>[{'Original_Sequence': 'TAGATGCCGATTCATTCTAGCTTCGACAACAGTCAGCGAAAACTCTACCCACAGTAAATATTTCAGCAAG',
  'New_Variant': 'TGAATGCCGATTCACTCTAGCTTCGACAACAGTCAACGAAAGCTCTACCCACAGTGAATTTTTCAACAAG',
  'Score': 0.6663332499583072,
  'Rank': 1,
  'Score_TRSp': 0.74,
  'Score_TRSpAvd': 0.6,
  'Proteins': Counter({'*MPIHSSFDNSQRKLYPQ*IFQQ': 8,
           '*MPVHSSFDNSQRKLYPQ*IFQQ': 5,
           '*MPIHSSFDNSQRKLSPQ*IFQQ': 4,
           '*MPIHSSFDNSQRELYPQ*IFQQ': 4,
           '*MPIHSSFDNSQRKLYPQ*IFPQ': 3,
           '*MPIHSSFANSQRKLYPQ*IFQQ': 3,
           '*MPLHSSFDNSQRKLYPQ*IFQQ': 3,
           '*MPIHSSFGNSQRKLYPQ*IFQQ': 3,
           '*MPIHSSFDNSHRKLYPQ*IFQQ': 3,
           '*MPIHSSFDTSQRKLYPQ*IFQQ': 2,
           '*MPIHSSFDNSQRKLYPQ*IFLQ': 2,
           '*MPIHSSFDSSQRKLYPQ*IFQQ': 2,
           '*MPIHSSFDNSQRQLYPQ*IFQQ': 2,
           '*MPIHSSFDNSQRKLYPQ*IFHQ': 2,
           '*MPIHSSFDNSQRRLYPQ*IFQQ': 2,
           'CMPLHSSFDYSLRKLYPQ*IFQQ': 1,
           '*MPIHSSFGNGRRKLYPQ*IFQQ': 1,
           '*MPIHSSFDNGQRELYPQ*IFLQ': 1,
           '*MPIHSSFANSLRELYPQ*IFQQ': 1,
           '*MPIHSSFDNSQRKLYPQ*IFQP': 1,
           '*MPIPSSFDNSQRKLCHQ*IFQQ': 1,
           '*MPIHSSFDYGLRKLYPQ*IFQQ': 1,
           '*MPIHSSFDHSQR*LYPQ*IFQQ': 1,
           '*MPIHSSFDTSQRELYPQ*IFQQ': 1,
           '*MPIHSSFDNSQRKLYPQ*VFQQ': 1,
           '*MPIQSSFVSSQRKLYPQ*IFQQ': 1,
           '*VPIHSSFDNSQRKLYPQ*IFQQ': 1,
           '*MPIHSSFDCGLRELYPQ*IFQQ': 1,
           '*MPIPSSFDSRQRKLYPP*IFQQ': 1,
           '*MPIHSSFDNSQRKLYPQ*IFQR': 1,
           '*MPIHSSFGSSQRKLYPQ*IFQQ': 1,
           '*MPIHSSFDNSLRELDPQ*IFQQ': 1,
           '*MPIHSSFDNSQRKLYPR*IFQQ': 1,
           '*MPLHSSFDHSQRKLYPQ*IFQQ': 1,
           'CMPIHSSFGGGQRKLSPQ*IFQQ': 1,
           '*MPIHSSFDYGQRKLYPQ*IFQQ': 1,
           '*MPIHSSFDNSQRELYPP*IFQQ': 1,
           '*MPIPSSFDNSQRKLYPQ*IFQQ': 1,
           '*MPIHYSFDNSQRQLYPQ*IFQR': 1,
           '*MPIPSSFDSSQRKLYPQ*IFQQ': 1,
           '*MPIHSSFGSSQR*LSPQCVFQQ': 1,
           '*MPIHSSFDKSQRKLYPR*VFQQ': 1,
           '*MPFLSSFDNSQRKLYPQ*IFQQ': 1,
           '*MPIHSSFGSSHRKLFPQ*IFQQ': 1,
           '*MPFHSSFANSQRKLYPQ*IFQQ': 1,
           '*MPIHSSFDNSQRKLYPQCIFQQ': 1,
           '*MPIHSSFDNSQRALYPQ*IFQQ': 1,
           '*MPIHSSFDASQRELYPQ*IFQQ': 1,
           '*MPIHSSFDYGQRGLSPQ*IFQQ': 1,
           '*MPIHSSFDNSPRKLYPQ*IFQQ': 1,
           '*VPIHSGFASSRRKLCPQ*IFQQ': 1,
           '*MPIHSSFDYRQRKLYPQ*IFQQ': 1,
           'WMPIHSGFDNSQRKLYPQ*IFQQ': 1,
           '*MPFHSSFDNSQRKLYPQ*IFQQ': 1,
           '*MPIHSSFDAGQRKLYPQ*IFQQ': 1,
           '*MPILSSFDDSQRKLYPQ*IFQQ': 1,
           '*MPIHSGFGSSQRKLYPQ*IFQQ': 1,
           '*MPIHSSFDNRQRKLYPQ*IFQQ': 1,
           '*MPIHSSFDASQRKLYPQ*IFQQ': 1,
           '*MPIHSSFDNSHRQLYPQ*IFQQ': 1,
           '*MPIHSSFDYSQRELYPQ*IFQQ': 1,
           '*LPIHSSFDNSQRKLYPQ*IFQQ': 1,
           '*MPIHSSFDGSQRKLSPQ*IFQQ': 1,
           '*MPIHSSFDNSLRKLYPQ*IFQQ': 1,
           '*MPIHSSFVSSQRKLYPQ*IFQQ': 1,
           '*MPIHSSFDSSRRKLSPQ*IFQQ': 1,
           '*MPIHSRFDNSQRKLYPQ*IFQQ': 1})},
 {'Original_Sequence': 'TAGATGCCGATTCATTCTAGCTTCGACAACAGTCAGCGAAAACTCTACCCACAGTAAATATTTCAGCAAG',
  'New_Variant': 'TGAATGCCGATTCACTCTAGCTTCGATAACAGTCAGCGAAAGCTCTACCCACAGTGAATTTTTCAACAAG',
  'Score': 0.6607571414672716,
  'Rank': 2,
  'Score_TRSp': 0.74,
  'Score_TRSpAvd': 0.59,
  'Proteins': Counter({'*MPIHSSFDNSQRKLYPQ*IFQQ': 11,
           '*MPIPSSFDNSQRKLYPQ*IFQQ': 5,
           '*MPIHSSFDNSQRKLSPQ*IFQQ': 3,
           '*MPIHSSFDNSQRKLYPQ*IFQH': 3,
           '*MPLHSSFDNSQRKLYPQ*IFQQ': 2,
           '*MPIHSSFDNSQRKLYPQ*IFPQ': 2,
           '*MPIHSSFDNSQRKLYPQ*IFHQ': 2,
           '*MPIHSSFDNSQRKLYPP*IFQQ': 2,
           '*MPIHSSFANSQRKLYPQ*IFQQ': 2,
           '*MPIHSSFDNSQRKLYPQ*IFQP': 2,
           '*VPIHSSFDNSQRKLYPQ*IFQQ': 2,
           '*MPVHSSFDNSQRKLYPQ*IFQQ': 2,
           '*MPIHSSFDTSQRKLYPQ*IFQQ': 2,
           '*MPIHSSFDNSRRKLYPQ*IFQP': 1,
           '*VPIHSSFDNSQRKLYPQ*IFQH': 1,
           '*MPIHSSFDRSQRKLYPQ*IFQQ': 1,
           '*MPIHSSFDYSQRKLYPQ*IFHQ': 1,
           '*MPIHSSFDNSQRMLCPQ*IFQQ': 1,
           '*MPIHSSFDDSQRELYPQ*IFQQ': 1,
           '*MPIHSSFDNSQRKLYPR*IFQR': 1,
           '*MPIHSSFDNSQRTLYPQ*IFQQ': 1,
           '*MPIPSSFDSSQRKLYPQ*IFQQ': 1,
           'WMPVLSSFDSSQRGLFPQ*IFQQ': 1,
           '*MPIHSSFDNSQRKLSPP*IFQQ': 1,
           '*MPIRSSFDSSQRKLYPQ*IFQQ': 1,
           '*MPIHSRFDNSQRKLYPQ*ICPQ': 1,
           '*MPIHSSFDYSQRKLDPQCIFQQ': 1,
           '*VPIHSSFDDSQRKLAPQ*IFQQ': 1,
           '*MPIHSSFDNSQRKLYPR*IFQQ': 1,
           '*MPIHSSFDKSQRKLYPQ*IFQH': 1,
           '*MPIHSSFDNSQRRLYPQ*IFQQ': 1,
           '*MPIHSSFDGSQRKLYPQ*IFQQ': 1,
           '*LPIHSSFDNSQRTLYPQ*IFQQ': 1,
           '*MPLHSSFDHSQRKLYPQ*IFQQ': 1,
           '*MPIHSSFDSSQRRLYPQ*IFQQ': 1,
           '*MPIPSSFDGSQRELYPQ*IFQQ': 1,
           '*MPIHSSFDYNRRKLYPQ*IFQQ': 1,
           '*MPIHSSFDTSQRTLYPQ*IFQQ': 1,
           '*MPIHSSFDGSQRELYPQ*IFRQ': 1,
           '*MPIHSSFDNSQRKL*PQCIFHQ': 1,
           '*MPIHSSFDNSRRKLYPQ*IFQQ': 1,
           'WLPIHSSFDNSQRKLYPQ*IFQQ': 1,
           '*MPFHSSFDNSQRKLYPQ*IFQQ': 1,
           '*MPIHSSFDYSQRKLYPQ*IFQQ': 1,
           '*MPFHSSFDGSQRKLCPQ*IFQQ': 1,
           '*VPVHSSFDNSQRKLYPQ*IFQQ': 1,
           '*MPIRSSFDNSQRKLYPQ*IFQQ': 1,
           '*MPILSSFGNSQRKLYPQ*IFQQ': 1,
           'CMPIHSSFDNSQRKLYPQ*IFQQ': 1,
           'WLPFRSSFDNSRRGLSPQWFFQQ': 1,
           '*MPIHSSFDYSQR*LYPQ*IFQQ': 1,
           '*MPILSSFDGGQRKLYPQ*IFQQ': 1,
           '*MPVHSSFDDSQRKLYPQ*IFQQ': 1,
           '*MPIHSSFDVSRRRLYHQCIFQQ': 1,
           '*MPIHSSFGLSQRKLSPQ*IFQQ': 1,
           '*MPIHSSFDVSQRKLYPQ*IFQQ': 1,
           '*MPIHSSFGNSQRKLYPQ*IFQQ': 1,
           '*MPIHSGFDYSQRKLYPQ*IFQQ': 1,
           '*MPIHSGFDYSQRKLSPQ*IFQQ': 1,
           '*MPIHSRFDNSQR*LYPQ*IFQQ': 1,
           '*MPIHSSFDNSQRKLYPQ*IFLQ': 1,
           '*MPIHSSFDNSQRKLSPR*IFQQ': 1,
           'WMPIHSSFDNSRRKLSPQWIFQQ': 1,
           '*LPIHSSFDSSQRKLYPQ*IFQQ': 1,
           '*MPIPSSFDNSQRQLYPQ*IFQQ': 1,
           '*MPIHSSFDDSQRKLCPQ*IFQQ': 1,
           '*MPIHSSFVNSQRKLYPQ*IFQQ': 1,
           '*MPIHSSFDNSQRKLYPQ*IFQR': 1,
           '*MPIHSSFDSSQRKLYPQ*IFQQ': 1,
           '*MPIHSSFDNSQRELSPQ*IFQQ': 1,
           '*MPIHSSFGYSQRKLYPQ*IFQQ': 1,
           '*MPIHSSFGNGQRKLSPQ*IFQQ': 1,
           '*MPIHSSFDNGLRKLFPQ*IFQQ': 1})},
 {'Original_Sequence': 'TAGATGCCGATTCATTCTAGCTTCGACAACAGTCAGCGAAAACTCTACCCACAGTAAATATTTCAGCAAG',
  'New_Variant': 'TGAATGCCGATTCACTCTAGCTTTGACAACAGTCAGCGAAAGCTCTACCCACAGTGAATTTTTCAACAAG',
  'Score': 0.6607571414672716,
  'Rank': 3,
  'Score_TRSp': 0.74,
  'Score_TRSpAvd': 0.59,
  'Proteins': Counter({'*MPIHSSFDNSQRKLYPQ*IFQQ': 13,
           '*MPIHSSFDSSQRKLYPQ*IFQQ': 6,
           '*MPIHSSFDNSQRKLYPQ*IFHQ': 5,
           '*VPIHSSFDNSQRKLYPQ*IFQQ': 5,
           '*MPIPSSFDNSQRKLYPQ*IFQQ': 4,
           '*MPVHSSFDNSQRKLYPQ*IFQQ': 3,
           '*MPIHSSFDNSQRKLSPQ*IFQQ': 3,
           '*MPIHSSFDDSQRKLYPQ*IFQQ': 3,
           '*MPIHSSFDASQRKLYPQ*IFQQ': 2,
           '*MPIHSSFDHSQRKLYPQ*IFQQ': 2,
           '*MPIHSSFDTSQRKLYPQ*IFQQ': 2,
           '*MPIHSSFDNSPRKLYPQ*IFQH': 1,
           '*MPIHSSFDNSQRKLDPQWIFHQ': 1,
           'CMPIHSSFDNSQRKLYPQ*IFQQ': 1,
           '*MPIRSSFGVSQRKLYAQ*IFQQ': 1,
           '*MPIPSSFGSSQRKLYPQ*IFQQ': 1,
           '*MPVPSSFGDSQRKLSPQ*IFQQ': 1,
           '*MPIRSSFDISQRELYPQ*IFQQ': 1,
           '*MPIHSSFDSSRRKLYPQ*IFQQ': 1,
           '*MPIHSSFDNRQRKLYPQ*IFQQ': 1,
           '*MPIHSGFGNSQRKLYPQ*IFQQ': 1,
           '*MPIPSGFDSSQRKLYPQ*IFQQ': 1,
           '*MPIHSSFDNSQRKLYPP*IFQQ': 1,
           '*MPIHSSFDPSQRKLYPQ*IFQQ': 1,
           '*MPIPSSFASSQRKLYPQCIFQQ': 1,
           '*MPIHSSFDNSPRKLYPQ*IFQQ': 1,
           '*MPIHSSFDSRQRKLYPQ*IFQQ': 1,
           '*MPIHSSFDNSQRKLAPQ*IFQQ': 1,
           '*LPIHSSFDNGQRKLYPQ*IFQQ': 1,
           '*MPIPFSFDVSQRKLYPQ*IFQQ': 1,
           '*MPVHSSFDNSQRELCPQ*IFQQ': 1,
           '*MPIHSSFDNSQRELSPQ*IFQQ': 1,
           'CLPIHSSFDNSQRKLYPQ*IFQQ': 1,
           '*MPIHSSFDNSQRGLCPQ*IFQQ': 1,
           '*MPIHSSFDNSQRKLYPQ*IFLQ': 1,
           '*MPIHSSFGYSQRKLYPQ*IFQQ': 1,
           '*MPFHSSFDYSQRKLYPQ*IFQQ': 1,
           'CMPIHSSFDNSPRKLYPQ*IFQQ': 1,
           '*MPIHSGFDGSQRKLYPQ*IFQQ': 1,
           '*MPIHSSFDNSRRKLYPQ*IFQQ': 1,
           '*MPVHSSFDSSQRKLYPQ*IFQQ': 1,
           '*MPIHSGFDSSQRGLVPQ*IFQQ': 1,
           '*MPIHSSFGNSRRKLYPQ*IFQQ': 1,
           '*MPIHSSFDYSQRKLYPQ*IFQQ': 1,
           '*LPIHSSFDNSQRKLYPQ*IFPQ': 1,
           '*MPLHSSFDNSQRKLYPQ*IFQQ': 1,
           '*MPLHSSFDNSHRKLYPQ*IFQQ': 1,
           '*MPIHSSFVNSRRKLSPQ*IFQQ': 1,
           '*VPIHSGFGDSQRKLYPQ*IFQQ': 1,
           '*MPVRSSFDNSQRKLSPQ*IFQQ': 1,
           '*MPIHSSFGNSQRKLYPQ*IFQQ': 1,
           '*MPIHSGFGYSQRELYPQ*IFQQ': 1,
           '*MPIHSSFANSQR*LSPQ*IFQQ': 1,
           '*MPILSSFDNGQRKLYPQ*IFQQ': 1,
           '*MPIHSSFVSSQRKLYPQ*IFQQ': 1,
           '*MPIHSSFDYSPRKLYPQ*IFQQ': 1,
           '*MPIHSSFVNSQRGLYPQ*IFQQ': 1,
           '*MPIHSSFDNSQRKLYPQ*LFQQ': 1,
           '*MPIHSSFDTSQRKLYPQ*IFQH': 1,
           '*MPIHSSFDNSQRKLYPQ*IFRQ': 1,
           '*MPIHSSFDYSQRELYPQ*IFQQ': 1,
           '*MPIHSSFDNSQRKLYPQ*IFPQ': 1,
           '*MPIHSSFRNSQRKLYPQ*IFQQ': 1})},
 {'Original_Sequence': 'TAGATGCCGATTCATTCTAGCTTCGACAACAGTCAGCGAAAACTCTACCCACAGTAAATATTTCAGCAAG',
  'New_Variant': 'TGAATGCCGATTCACAGCAGCTTCGACAACAGTCAGCGAAAACTCTACCCGCAGTGAATCTTTCAACAAG',
  'Score': 0.6587867636800242,
  'Rank': 4,
  'Score_TRSp': 0.7,
  'Score_TRSpAvd': 0.62,
  'Proteins': Counter({'*MPIHSSFDNSQRKLYPQ*IFQQ': 8,
           '*MPIPSSFDNSQRKLYPQ*IFQQ': 4,
           '*MPIRSSFDNSQRKLYPQ*IFQQ': 3,
           '*MPIHSSFDNSPRKLYPQ*IFQQ': 3,
           '*MPIHSSFDNSQRKLYPQ*IFHQ': 3,
           '*MPIHSSFDNSQRNLYPQ*IFQQ': 2,
           '*MPIHSSFDNSQRTLYPQ*IFQQ': 2,
           '*MPIHSSFANSQRKLYPQ*IFQQ': 2,
           '*MPVHSSFDNSQRKLYPQ*IFQQ': 2,
           '*MPIHSSFDNSQRKLSPQ*IFQQ': 2,
           '*MPIRSGFDLSQRKLYPQWIFKQ': 1,
           '*MPIHSSFDNSQRTLSPQ*IFQQ': 1,
           '*MPIHSSFDNSQRKLYPQWIFQQ': 1,
           '*MPIHSSFDNSRRELCPQ*IFQQ': 1,
           '*MPLHSSFDNSQRKLYPQ*IFQQ': 1,
           '*MPIHSGFDDSQRKLYPQ*IFQQ': 1,
           '*MPIHSRFDNSQRKLYPQ*LFQQ': 1,
           '*MPIHSSFDNSQRKLYPQ*IFPQ': 1,
           '*VPIRSSFDNSQRKLYPQ*IFQQ': 1,
           '*MPIRSSFDNSQRSLSPQ*IFQQ': 1,
           '*MPIHSSFDNSQRKLYPQCIFQH': 1,
           '*VPIRSSFDSGRRILCPQ*IFKR': 1,
           '*MPIPSSFDNSQRKLYPQ*IFQP': 1,
           '*MPIHSSFDTSQRKLYPQ*IFHQ': 1,
           '*MPIHGSFDYSQRKLYPQ*IFHQ': 1,
           '*MPICSGFVNSQRKLYPQ*IFQQ': 1,
           '*MPIRCSVGNSQRGLYPQ*IFQQ': 1,
           '*MPIHSSFDTSQRKLYPQ*IFQH': 1,
           '*MPIHSSFDSSPRKLYPQ*IFHQ': 1,
           '*MPIHSSFGLSQRKLYPQ*IFQQ': 1,
           '*MPLHRSFDNSQRKLYPQ*IFQQ': 1,
           '*MPIHSSFDNGQRKLYPQ*IFQQ': 1,
           '*MPIHSSFDTSQRKLYPQCIFQQ': 1,
           'CMPIHSSFANSQRKLYPQ*IFQQ': 1,
           '*MPIHSSFVNSQRKLYPQ*IFQQ': 1,
           '*MPIRGSFDNSRRKLYPQ*IFQQ': 1,
           '*MPIHSSFGASQRKLYPQ*IFQQ': 1,
           '*MPLHSSFDNSQRKLSPQ*IFQQ': 1,
           '*MPIHGSFDDSQRKLYPQ*IFQQ': 1,
           '*MPIHSSFDNSRRLLYPQ*IFQQ': 1,
           '*MPIRSGFDNSQRKLYPQ*IFQQ': 1,
           '*MPIHSSFDNSQRKLYPQ*IFLQ': 1,
           '*MPIHSGFDFSQRKLYPQ*IFQQ': 1,
           '*MPIRSGFANSQRLLYPQ*IFQQ': 1,
           '*MPIHSSFDSRQRILYPQ*IFQQ': 1,
           '*MPIGSSFDCSQRKLYPQ*IFQQ': 1,
           '*LPIHSSFDSSQRKLYPQ*IFQQ': 1,
           '*MPIHSSFGNSQRKLYPQ*IFQQ': 1,
           'CMPIHRSFDNSQRKLYPQ*IFQQ': 1,
           '*MPIRSGFDSSQRGLSPQ*IFQQ': 1,
           '*MPIHSSFDISRRKLYPQ*IFQQ': 1,
           '*MPIHSSFDHSQRKLYPQ*IFQQ': 1,
           '*MPILSSFDASQRKLSPQWIFQH': 1,
           '*MPIFSSFDFSQRILYPQ*IFQQ': 1,
           '*MPIHSSFDNRQRKLYPQ*IFQQ': 1,
           '*MPIHSSFDNSQRQLYPQ*IFQQ': 1,
           '*MPIHSSFDNGQRLLFPQ*IFQQ': 1,
           '*MPIHSSFDSSQRGLYPQ*IFQQ': 1,
           '*VPFHSSFDSSQRKLYPQ*IFQQ': 1,
           '*MPIHSSFDNSQRKLYPQ*IFRQ': 1,
           '*LPIHSSFDNSQRKLYPQ*IFQQ': 1,
           '*MPIHSSFDYRQRKLYPQ*IFQQ': 1,
           '*MPIHRSFDNSQRKLYPQ*IFHQ': 1,
           '*MPIHSSFEYSQRPLYPQ*LFQQ': 1,
           'CMPIRSGFGDSQRKLYPQ*IFQQ': 1,
           'CLPFRSSFDNGQRKLYPQ*IFQQ': 1,
           '*MPIHSSFDNGRRKLYPQ*IFQQ': 1,
           '*MPIHSSFGNSQRELDPQ*IFQQ': 1,
           '*MPIHGSFDNGQRALSPRCIFQQ': 1,
           '*MPIHSSFDYSQRLLYPQ*IFQQ': 1,
           '*MPVHSRFDNSQRELYSR*IFQQ': 1,
           '*MPINSSFDNRQRKLYPQ*IFQQ': 1,
           '*MPIHSGFDYSQRNLYPQ*IFQQ': 1,
           '*MPIPSSFDNSQRSLSPQ*IFQQ': 1,
           '*MPIHSSFDNSQRRLYPQ*IFQQ': 1,
           '*MPIRSSVDNSQRELYPQ*IFQQ': 1,
           '*MPIHSSFDNSQRKLYPQ*IFPH': 1,
           '*MPIHSSFDFSQRLLYPQ*IFQQ': 1,
           '*MPIHGSFDNSQRLLYPQ*IFQQ': 1})},
 {'Original_Sequence': 'TAGATGCCGATTCATTCTAGCTTCGACAACAGTCAGCGAAAACTCTACCCACAGTAAATATTTCAGCAAG',
  'New_Variant': 'TGAATGCCGATTCACAGCAGCTTCGACAACAGTCAGCGAAAACTCTATCCACAGTGAATCTTTCAACAAG',
  'Score': 0.6587867636800242,
  'Rank': 5,
  'Score_TRSp': 0.7,
  'Score_TRSpAvd': 0.62,
  'Proteins': Counter({'*MPIHSSFDNSQRKLYPQ*IFQQ': 7,
           '*MPIHSSFDNRQRKLYPQ*IFQQ': 4,
           '*MPIHSSFANSQRKLYPQ*IFQQ': 3,
           '*MPIHSRFDNSQRKLYPQ*IFQQ': 3,
           '*MPIHSSFDASQRKLYPQ*IFQQ': 2,
           '*MPIHSSFDYSQRKLYPQ*IFQQ': 2,
           '*MPLHSSFDNSQRKLYPQ*IFQQ': 2,
           '*MPIHSSFDTSQRKLYPQ*IFQQ': 2,
           '*MPIRSSFDSSQRELYPQ*IFQQ': 2,
           '*MPIHSSFDNSQRKLYPQ*IFQH': 2,
           '*VPIHSSFDNSQRKLYPQ*IFQQ': 2,
           '*MPVHSSFDNSQRKLYPQ*IFQQ': 2,
           '*MPIHSSFDNSQRKLYPP*IFQQ': 2,
           '*MPIHSSFDHSQRKLYPQ*IFQQ': 2,
           '*MPIHGSFDNSQRKLYPQ*IFQQ': 2,
           '*MPIPSSFDNSQRKLYPQ*IFQQ': 2,
           '*MPIHSSFDSSQRQLYPQ*IFQQ': 1,
           '*MPIHSSFDNSQRKLYPQCIFQQ': 1,
           '*VPIRSSFDNSQRKLYPQ*IFQQ': 1,
           '*MPIHSSFDNSQRPLYPQ*IFQQ': 1,
           '*LPLHSSFDNSQRKLYPQ*IFQQ': 1,
           '*MPIHSGFGHRQRKLYPQ*IFQH': 1,
           '*MPFHSSFGSSQRKLYPQ*IFQQ': 1,
           '*MQIRSSFDTSQRKLYPQ*IFQQ': 1,
           '*MPIHGSFDNSQRELYPQ*IFQQ': 1,
           '*MPIPSSFDSSQRILYPQ*IFQQ': 1,
           '*MPVLSSFVNSRRELYPQ*IFQQ': 1,
           '*MPIHSSFDHSQRKLYPQCIFQQ': 1,
           '*MPIHSSFDDSQRILGPQ*IFQQ': 1,
           '*MPIHSSFDHSQRILYPQ*IFQQ': 1,
           '*MPICSGFDDSQRKLYPQ*IFQQ': 1,
           '*MPIPGSFDNSQRKLYPQ*IFQQ': 1,
           '*MPIHSSFDNSQRKLYPQ*IFQP': 1,
           '*MPIHSSFADSQRALYPQ*IFQQ': 1,
           '*MPIRSSFVNSQRKLYPQ*IFQQ': 1,
           '*MPIRSGFDNSQRKLYPQ*IFQQ': 1,
           '*MPIHSSFDFSQR*LYPP*IFQQ': 1,
           '*MPFHSSFDNSQRKLYPQ*IFQQ': 1,
           '*MPIHSSFGSSQRKLYPQ*IFQQ': 1,
           '*MPLHSGFDNSQRKLYPQ*IFQQ': 1,
           '*MPIHSSFDNSQRELYPQ*IFQQ': 1,
           '*MPIHSSFDNSQRKLYPQCIFQH': 1,
           '*MPIHSSFDNSQRQLYPQ*IFQQ': 1,
           '*MPIHCGFVNSQRKLYPQ*IFQQ': 1,
           '*LPIHSSFDNSQRKLYPQ*IFQQ': 1,
           '*MPIRSSFDNSQRKLYPR*IFQQ': 1,
           '*MPIHRSFDNSQRKLYPQ*IFQQ': 1,
           '*MPIHSRFDNSQRKLYPQ*LFQQ': 1,
           '*MPIHSGFDNSQRKLYPQ*IFQQ': 1,
           '*MPLHSSFDNSQRKLYPQ*IFQR': 1,
           '*MPIHSSFDHGQRKLYPQ*IFQQ': 1,
           '*MPIRSSFDYSQRKLCPQ*IFQQ': 1,
           '*MPIHGSFDASQRKLYPQ*IFQQ': 1,
           '*MPCHCGFDVCQRKLYPQ*IFQQ': 1,
           'WLPVHSSFDDGQRKLYPQ*IFQQ': 1,
           '*MPIHSSFDHSQRNLYPQ*IFQQ': 1,
           '*MPIHSSFDNSQRKLYPQ*IFHQ': 1,
           '*MPIHSSFDNSQRNLVPQ*IFKQ': 1,
           '*MPIRSSFANSQRKLYPQ*IFQQ': 1,
           '*MPIHSSFDNSRRKLYPQ*IFQQ': 1,
           'CMPVHSSFDNSQRKLYPQ*IFQQ': 1,
           '*MPIRSSFDGSRRGLYPQ*IFQQ': 1,
           '*MPIHSSFDNSQRILYPQ*IFQQ': 1,
           'WLPIHSGFANSQRKLYPQ*IFQQ': 1,
           '*MPLHSSFDNSQRKLCPQ*IFQQ': 1,
           '*MPIPSSFDNSQRKLYPQ*ILQQ': 1,
           '*MPFPGCFDCGPRKLYPQ*IFQQ': 1,
           '*MPIRSSFDYSQRILYPR*IFQQ': 1,
           '*MPIHSSFDGSQRKLFPQ*IFQQ': 1,
           '*MPIHSSFDSSQRKLYPQ*IFQQ': 1,
           '*MPIHGSFDNSQRKLYPQ*IFHQ': 1,
           '*MPIHSSFDNSQRRLYPQ*IFQQ': 1,
           '*MPVPGSFDNSQRLLYPQ*IFLQ': 1,
           'CMPIHSSFDNSQRKLYPQ*IFQQ': 1,
           '*MPIRSSFDSSQRKLYPQ*IFQQ': 1})}]</code></pre>
</div>
</div>
<hr>
<p><a href="https://github.com/dbikard/dgrec/blob/main/dgrec/lstm.py#L599" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="compute_likelihood" class="level3">
<h3 class="anchored" data-anchor-id="compute_likelihood">compute_likelihood</h3>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb33"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> compute_likelihood(</span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a>    TR, <span class="co"># Reference/template sequence.</span></span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a>    VR, <span class="co"># Variant sequence to score.</span></span>
<span id="cb33-5"><a href="#cb33-5" aria-hidden="true" tabindex="-1"></a>): <span class="co"># Log-likelihood of VR given TR.</span></span>
<span id="cb33-6"><a href="#cb33-6" aria-hidden="true" tabindex="-1"></a></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p><em>Compute the log-likelihood of generating a variant sequence (VR)</em> given a template/reference sequence (TR).</p>
<div id="6375bf93-05c5-44a5-9757-401edb33faaa" class="cell" data-execution_count="12">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb34"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>TR<span class="op">=</span><span class="st">'CGTAAACCGGACCTAGTTTAGTTCTTAGACCAAGGTACATATCCCCGTAACATAAGACGCGACTGGGCCC'</span></span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>VR<span class="op">=</span><span class="st">'CGTAAACCGGACCTAGTTTAGTTCTTAGACCAAGGTACATATCCCCGTAACATAAGACGCGACTGGGCCC'</span></span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a>compute_likelihood(TR, VR)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display" data-execution_count="12">
<pre><code>-4.534628229418797</code></pre>
</div>
</div>
<hr>
<p><a href="https://github.com/dbikard/dgrec/blob/main/dgrec/lstm.py#L746" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="compute_likelihood_list" class="level3">
<h3 class="anchored" data-anchor-id="compute_likelihood_list">compute_likelihood_list</h3>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb36"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> compute_likelihood_list(</span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a>    TR_list, <span class="co"># Reference/template sequences.</span></span>
<span id="cb36-4"><a href="#cb36-4" aria-hidden="true" tabindex="-1"></a>    VR_list, <span class="co"># Variant sequences.</span></span>
<span id="cb36-5"><a href="#cb36-5" aria-hidden="true" tabindex="-1"></a>): <span class="co"># Log-likelihoods in the same order as input.</span></span>
<span id="cb36-6"><a href="#cb36-6" aria-hidden="true" tabindex="-1"></a></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p><em>Compute log-likelihoods for (TR, VR) pairs.</em></p>
<div id="86b742be-5dd7-416b-9e8c-3b7ea1ed7485" class="cell" data-execution_count="15">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb37"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>TR_list<span class="op">=</span>[<span class="st">'CGTAAACCGGACCTAGTTTAGTTCTTAGACCAAGGTACATATCCCCGTAACATAAGACGCGACTGGGCCC'</span>]<span class="op">*</span><span class="dv">2</span></span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>VR_list<span class="op">=</span>[<span class="st">'CGTAAACCGGACCTAGTTTAGTTCTTAGACCAAGGTACATATCCCCGTAACATAAGACGCGACTGGGCCC'</span>,<span class="st">'CGTAAACCGGACCTAGTTTAGTTCTTAGACCAAGGTACATATCCCCGTATCATAAGACGCGACTGGGCCC'</span>]</span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a>compute_likelihood_list(TR_list, VR_list)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display" data-execution_count="15">
<pre><code>[-4.535991900941974, -8.759227246265613]</code></pre>
</div>
</div>
<hr>
<p><a href="https://github.com/dbikard/dgrec/blob/main/dgrec/lstm.py#L793" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="compute_likelihood_matrix" class="level3">
<h3 class="anchored" data-anchor-id="compute_likelihood_matrix">compute_likelihood_matrix</h3>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb39"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> compute_likelihood_matrix(</span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a>    TR_list, <span class="co"># List of reference/template sequences.</span></span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a>    VR_list, <span class="co"># List of variant sequences.</span></span>
<span id="cb39-5"><a href="#cb39-5" aria-hidden="true" tabindex="-1"></a>    batch_size:<span class="bu">int</span><span class="op">=</span><span class="dv">64</span>, <span class="co"># (Currently unused) Intended batch size for future optimization.</span></span>
<span id="cb39-6"><a href="#cb39-6" aria-hidden="true" tabindex="-1"></a>): <span class="co"># Log-likelihood matrix of shape (len(TR_list), len(VR_list)).</span></span>
<span id="cb39-7"><a href="#cb39-7" aria-hidden="true" tabindex="-1"></a></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p><em>Compute a matrix of log-likelihoods where each entry (i, j)</em> corresponds to the log-likelihood of generating VR_list[j] from TR_list[i].</p>
<p>Sequences with mismatched lengths are assigned -inf.</p>
<div id="60af2646-55af-46ba-8a97-a029f0be16ed" class="cell" data-execution_count="17">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb40"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>TR_list<span class="op">=</span>[<span class="st">'CGTAAACCGGACCTAGTTTAGTTCTTAGACCAAGGTACATATCCCCGTAACATAAGACGCGACTGGGCCC'</span>,<span class="st">'CGTAAACCGGACCTAGTTTAGTTCTTAGACCAAGGTACATATCCCCGTATCATAAGACGCGACTGGGCCC'</span>]</span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a>VR_list<span class="op">=</span>[<span class="st">'CGTAAACCGGACCTAGTTTAGTTCTTAGACCAAGGTACATATCCCCGTAACATAAGACGCGACTGGGCCC'</span>,<span class="st">'CGTAAACCGGACCTAGTTTAGTTCTTAGACCAAGGTACATATCCCCGTATCATAAGACGCGACTGGGCCC'</span>]</span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a>compute_likelihood_matrix(TR_list, VR_list)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display" data-execution_count="17">
<pre><code>[[-4.535991900941974, -8.759227246265613],
 [-16.457268648590745, -5.387971227326468]]</code></pre>
</div>
</div>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
      const outerScaffold = trigger.parentElement.cloneNode(true);
      const codeEl = outerScaffold.querySelector('code');
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp("https:\/\/dbikard\.github\.io\/dgrec");
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->




<footer class="footer"><div class="nav-footer"><div class="nav-footer-center"><div class="toc-actions d-sm-block d-md-none"><ul><li><a href="https://github.com/dbikard/dgrec/issues/new" class="toc-action"><i class="bi bi-github"></i>Report an issue</a></li></ul></div></div></div></footer></body></html>