<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.7.32">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>utils – dgrec</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<script src="../site_libs/quarto-html/quarto.js" type="module"></script>
<script src="../site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-37eea08aefeeee20ff55810ff984fec1.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap-2c2b2e643518224cf2d75a643cacf640.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="../styles.css">
<meta property="og:title" content="utils – dgrec">
<meta property="og:description" content="Analysing DGRec data">
<meta property="og:site_name" content="dgrec">
<meta name="twitter:title" content="utils – dgrec">
<meta name="twitter:description" content="Analysing DGRec data">
<meta name="twitter:card" content="summary">
</head>

<body class="nav-sidebar floating nav-fixed quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">dgrec</span>
    </a>
  </div>
        <div class="quarto-navbar-tools tools-end">
</div>
          <div id="quarto-search" class="" title="Search"></div>
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../API/genotypes.html">API</a></li><li class="breadcrumb-item"><a href="../API/utils.html">utils</a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">dgrec</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true">
 <span class="menu-text">API</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../API/genotypes.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">genotypes</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../API/genotypes_paired.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">genotypes_paired</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../API/plotting.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">plotting</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../API/analysis.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">analysis</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../API/utils.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">utils</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../API/cli.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Command line interface</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true">
 <span class="menu-text">DGRec data analysis</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../DGRec data analysis/mutation_distribution.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Mutations distribution</span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#align2mut" id="toc-align2mut" class="nav-link active" data-scroll-target="#align2mut">align2mut</a></li>
  <li><a href="#mut_rix" id="toc-mut_rix" class="nav-link" data-scroll-target="#mut_rix">mut_rix</a></li>
  <li><a href="#get_mutations" id="toc-get_mutations" class="nav-link" data-scroll-target="#get_mutations">get_mutations</a></li>
  <li><a href="#get_mutations_noalign" id="toc-get_mutations_noalign" class="nav-link" data-scroll-target="#get_mutations_noalign">get_mutations_noalign</a></li>
  <li><a href="#mut_to_str" id="toc-mut_to_str" class="nav-link" data-scroll-target="#mut_to_str">mut_to_str</a></li>
  <li><a href="#str_to_mut" id="toc-str_to_mut" class="nav-link" data-scroll-target="#str_to_mut">str_to_mut</a></li>
  <li><a href="#genstr_to_seq" id="toc-genstr_to_seq" class="nav-link" data-scroll-target="#genstr_to_seq">genstr_to_seq</a></li>
  <li><a href="#reverse_complement" id="toc-reverse_complement" class="nav-link" data-scroll-target="#reverse_complement">reverse_complement</a></li>
  <li><a href="#get_prot_mut" id="toc-get_prot_mut" class="nav-link" data-scroll-target="#get_prot_mut">get_prot_mut</a></li>
  <li><a href="#parse_genotypes" id="toc-parse_genotypes" class="nav-link" data-scroll-target="#parse_genotypes">parse_genotypes</a></li>
  <li><a href="#get_aa_mut_list" id="toc-get_aa_mut_list" class="nav-link" data-scroll-target="#get_aa_mut_list">get_aa_mut_list</a></li>
  <li><a href="#downsample_fastq_gz" id="toc-downsample_fastq_gz" class="nav-link" data-scroll-target="#downsample_fastq_gz">downsample_fastq_gz</a></li>
  <li><a href="#get_basename_without_extension" id="toc-get_basename_without_extension" class="nav-link" data-scroll-target="#get_basename_without_extension">get_basename_without_extension</a></li>
  <li><a href="#pickle_save" id="toc-pickle_save" class="nav-link" data-scroll-target="#pickle_save">pickle_save</a></li>
  <li><a href="#pickle_load" id="toc-pickle_load" class="nav-link" data-scroll-target="#pickle_load">pickle_load</a></li>
  <li><a href="#make_dgr_oligos" id="toc-make_dgr_oligos" class="nav-link" data-scroll-target="#make_dgr_oligos">make_dgr_oligos</a></li>
  <li><a href="#reverse_comp_geno_list" id="toc-reverse_comp_geno_list" class="nav-link" data-scroll-target="#reverse_comp_geno_list">reverse_comp_geno_list</a></li>
  <li><a href="#remove_position" id="toc-remove_position" class="nav-link" data-scroll-target="#remove_position">remove_position</a></li>
  <li><a href="#remove_position_list" id="toc-remove_position_list" class="nav-link" data-scroll-target="#remove_position_list">remove_position_list</a></li>
  </ul>
<div class="toc-actions"><ul><li><a href="https://github.com/dbikard/dgrec/issues/new" class="toc-action"><i class="bi bi-github"></i>Report an issue</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../API/genotypes.html">API</a></li><li class="breadcrumb-item"><a href="../API/utils.html">utils</a></li></ol></nav>
<div class="quarto-title">
<h1 class="title">utils</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<!-- WARNING: THIS FILE WAS AUTOGENERATED! DO NOT EDIT! -->
<hr>
<p><a href="https://github.com/dbikard/dgrec/blob/main/dgrec/utils.py#L21" target="_blank" style="float:right; font-size:smaller">source</a></p>
<section id="align2mut" class="level3">
<h3 class="anchored" data-anchor-id="align2mut">align2mut</h3>
<blockquote class="blockquote">
<pre><code> align2mut (align)</code></pre>
</blockquote>
<p><em>Converts a sequence alignment result from Bio.pairwise2.Align.globalms into a list of mutations. Positions are those of the alignment.</em></p>
<div id="cell-3" class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>seqA <span class="op">=</span> <span class="st">"ATCCCGGCAGC"</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>seqB <span class="op">=</span> <span class="st">"ATCCACGGTCAGC"</span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>align<span class="op">=</span>pairwise2.align.globalms(seqA,seqB, <span class="dv">2</span>, <span class="op">-</span><span class="dv">1</span>, <span class="op">-</span><span class="dv">1</span>, <span class="op">-</span><span class="fl">.5</span>, one_alignment_only<span class="op">=</span><span class="va">True</span>)[<span class="dv">0</span>]</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>align2mut(align)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>[('-', 4, 'A'), ('-', 8, 'T')]</code></pre>
</div>
</div>
<hr>
<p><a href="https://github.com/dbikard/dgrec/blob/main/dgrec/utils.py#L32" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="mut_rix" class="level3">
<h3 class="anchored" data-anchor-id="mut_rix">mut_rix</h3>
<blockquote class="blockquote">
<pre><code> mut_rix (mutations)</code></pre>
</blockquote>
<p><em>Reindexes the positions of the mutations to go from their position in the sequence alignment to their position in the original sequence.</em></p>
<div id="cell-5" class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>seqA <span class="op">=</span> <span class="st">"ATCCCGGCAGC"</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>seqB <span class="op">=</span> <span class="st">"ATCCACGGTCAGC"</span></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>align<span class="op">=</span>pairwise2.align.globalms(seqA,seqB, <span class="dv">2</span>, <span class="op">-</span><span class="dv">1</span>, <span class="op">-</span><span class="dv">1</span>, <span class="op">-</span><span class="fl">.5</span>, one_alignment_only<span class="op">=</span><span class="va">True</span>)[<span class="dv">0</span>]</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(format_alignment(<span class="op">*</span>align))</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>mutations<span class="op">=</span>align2mut(align) </span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Output of align2mut:"</span>)</span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(mutations)</span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Output of mut_rix:"</span>)</span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(mut_rix(mutations))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>ATCC-CGG-CAGC
|||| ||| ||||
ATCCACGGTCAGC
  Score=20

Output of align2mut:
[('-', 4, 'A'), ('-', 8, 'T')]
Output of mut_rix:
[('-', 4, 'A'), ('-', 7, 'T')]</code></pre>
</div>
</div>
<hr>
<p><a href="https://github.com/dbikard/dgrec/blob/main/dgrec/utils.py#L46" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="get_mutations" class="level3">
<h3 class="anchored" data-anchor-id="get_mutations">get_mutations</h3>
<blockquote class="blockquote">
<pre><code> get_mutations (seqA, seqB, match=2, mismatch=-1, gap_open=-1,
                gap_extend=-0.5)</code></pre>
</blockquote>
<p><em>Aligns two sequences and returns a genotype string. The string is a comma separated list of mutations.</em></p>
<div id="cell-7" class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>seqA <span class="op">=</span> <span class="st">"ATCCGGCAGCAGGTCGTGAGC"</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>seqB <span class="op">=</span> <span class="st">"ATCCACGGTCAGCACGTCGTGGC"</span></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>align<span class="op">=</span>pairwise2.align.globalms(seqA,seqB, <span class="dv">2</span>, <span class="op">-</span><span class="dv">1</span>, <span class="op">-</span><span class="dv">1</span>, <span class="op">-</span><span class="fl">.5</span>, one_alignment_only<span class="op">=</span><span class="va">True</span>)[<span class="dv">0</span>]</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(format_alignment(<span class="op">*</span>align))</span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>get_mutations(seqA,seqB)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>ATC--CGG-CAGCAGGTCGTGAGC
|||  ||| |||||.|||||| ||
ATCCACGGTCAGCACGTCGTG-GC
  Score=33.5
</code></pre>
</div>
<div class="cell-output cell-output-display">
<pre><code>[('-', 3, 'C'), ('-', 3, 'A'), ('-', 6, 'T'), ('G', 11, 'C'), ('A', 18, '-')]</code></pre>
</div>
</div>
<hr>
<p><a href="https://github.com/dbikard/dgrec/blob/main/dgrec/utils.py#L56" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="get_mutations_noalign" class="level3">
<h3 class="anchored" data-anchor-id="get_mutations_noalign">get_mutations_noalign</h3>
<blockquote class="blockquote">
<pre><code> get_mutations_noalign (seqA, seqB)</code></pre>
</blockquote>
<p><em>Returns a genotype string. The string is a comma separated list of mutations.</em></p>
<div id="cell-9" class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>get_mutations_noalign(<span class="st">"AGCTATGG"</span>,<span class="st">"AGCTCTGG"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>[('A', '4', 'C')]</code></pre>
</div>
</div>
<hr>
<p><a href="https://github.com/dbikard/dgrec/blob/main/dgrec/utils.py#L68" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="mut_to_str" class="level3">
<h3 class="anchored" data-anchor-id="mut_to_str">mut_to_str</h3>
<blockquote class="blockquote">
<pre><code> mut_to_str (mutations:list)</code></pre>
</blockquote>
<p><em>Converts list of mutations to a comma separated string</em></p>
<div id="cell-11" class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>mut_to_str([(<span class="st">'-'</span>, <span class="dv">3</span>, <span class="st">'C'</span>), (<span class="st">'-'</span>, <span class="dv">3</span>, <span class="st">'A'</span>), (<span class="st">'-'</span>, <span class="dv">6</span>, <span class="st">'T'</span>), (<span class="st">'G'</span>, <span class="dv">11</span>, <span class="st">'C'</span>), (<span class="st">'A'</span>, <span class="dv">18</span>, <span class="st">'-'</span>)])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>'-3C,-3A,-6T,G11C,A18-'</code></pre>
</div>
</div>
<hr>
<p><a href="https://github.com/dbikard/dgrec/blob/main/dgrec/utils.py#L75" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="str_to_mut" class="level3">
<h3 class="anchored" data-anchor-id="str_to_mut">str_to_mut</h3>
<blockquote class="blockquote">
<pre><code> str_to_mut (gen:str)</code></pre>
</blockquote>
<p><em>Converts genotype string to a list of mutations</em></p>
<div id="cell-13" class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="cf">assert</span>(str_to_mut(<span class="st">''</span>)<span class="op">==</span>[])</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>str_to_mut(<span class="st">'-4A,-7T,G12C'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>[['-', 4, 'A'], ['-', 7, 'T'], ['G', 12, 'C']]</code></pre>
</div>
</div>
<hr>
<p><a href="https://github.com/dbikard/dgrec/blob/main/dgrec/utils.py#L92" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="genstr_to_seq" class="level3">
<h3 class="anchored" data-anchor-id="genstr_to_seq">genstr_to_seq</h3>
<blockquote class="blockquote">
<pre><code> genstr_to_seq (genstr, refseq)</code></pre>
</blockquote>
<div id="cell-15" class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>genstr<span class="op">=</span><span class="st">'-3C,-3A,-6T,G11C,A18-'</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>genstr_to_seq(genstr,seqA)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>'ATCCACGGTCAGCACGTCGTGGC'</code></pre>
</div>
</div>
<div id="cell-16" class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>s<span class="op">=</span><span class="st">"AGCGAGC"</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="bu">len</span>(s)<span class="op">%</span><span class="dv">3</span>)</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>s[:<span class="op">-</span>(<span class="bu">len</span>(s)<span class="op">%</span><span class="dv">3</span>)]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>1</code></pre>
</div>
<div class="cell-output cell-output-display">
<pre><code>'AGCGAG'</code></pre>
</div>
</div>
<hr>
<p><a href="https://github.com/dbikard/dgrec/blob/main/dgrec/utils.py#L113" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="reverse_complement" class="level3">
<h3 class="anchored" data-anchor-id="reverse_complement">reverse_complement</h3>
<blockquote class="blockquote">
<pre><code> reverse_complement (dna)</code></pre>
</blockquote>
<hr>
<p><a href="https://github.com/dbikard/dgrec/blob/main/dgrec/utils.py#L129" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="get_prot_mut" class="level3">
<h3 class="anchored" data-anchor-id="get_prot_mut">get_prot_mut</h3>
<blockquote class="blockquote">
<pre><code> get_prot_mut (genstr, refseq, frame=0, ori=1)</code></pre>
</blockquote>
<div id="cell-19" class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>seqA <span class="op">=</span> <span class="st">"AGCGCTATGCTGCGCGCGTACTGCCGCTAGCTATGCTCAGGCCGATATATGCGAGCA"</span></span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>seqB <span class="op">=</span> <span class="st">"AGCGCTATGCTGCGCGCGAAAACCCGCTAGCTATGCTCAGGTCGATATATGCGAGCA"</span></span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>genstr<span class="op">=</span>mut_to_str(get_mutations(seqA,seqB,gap_open<span class="op">=-</span><span class="dv">5</span>))</span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(genstr)</span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a>get_prot_mut(genstr,seqA,frame<span class="op">=</span><span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>T18A,C20A,T21A,G22C,C41T</code></pre>
</div>
<div class="cell-output cell-output-display">
<pre><code>'T6K,A7P,A13V'</code></pre>
</div>
</div>
<div id="cell-20" class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>genstr<span class="op">=</span><span class="st">'A61G,-63T,A79T'</span></span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>refseq<span class="op">=</span><span class="st">'CGCCTTGGTAGCCATCTTCAGTTCCAGTGTTTGCTTCAAATACTAAGTATTTGTGGCCTTTATCTTCTACGTAGTGAGGATCTCTCAGCGTATGGTTGTCGCCTGAGCTGTAGTTGCCTTCATCGATGAACTGCTGTAC'</span></span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a>ori<span class="op">=-</span><span class="dv">1</span></span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a>frame<span class="op">=</span><span class="dv">0</span></span>
<span id="cb31-5"><a href="#cb31-5" aria-hidden="true" tabindex="-1"></a>get_prot_mut(genstr,refseq,frame<span class="op">=</span>frame,ori<span class="op">=</span>ori)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>'D19E,D25E,K26Q,G27R,H28P,K29Q,Y30I,V32S,F33I,E34*,A35S,N36K,T37H,G38W,T39N,E40*,D41R,G42W,Y43L,Q44P,G45R'</code></pre>
</div>
</div>
<div id="cell-21" class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>genstr<span class="op">=</span><span class="st">'A61G,A79T'</span></span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>refseq<span class="op">=</span><span class="st">'CGCCTTGGTAGCCATCTTCAGTTCCAGTGTTTGCTTCAAATACTAAGTATTTGTGGCCTTTATCTTCTACGTAGTGAGGATCTCTCAGCGTATGGTTGTCGCCTGAGCTGTAGTTGCCTTCATCGATGAACTGCTGTAC'</span></span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a>ori<span class="op">=-</span><span class="dv">1</span></span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a>frame<span class="op">=</span><span class="dv">0</span></span>
<span id="cb33-5"><a href="#cb33-5" aria-hidden="true" tabindex="-1"></a>get_prot_mut(genstr,refseq,frame<span class="op">=</span>frame,ori<span class="op">=</span>ori)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>'D19E'</code></pre>
</div>
</div>
<hr>
<p><a href="https://github.com/dbikard/dgrec/blob/main/dgrec/utils.py#L156" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="parse_genotypes" class="level3">
<h3 class="anchored" data-anchor-id="parse_genotypes">parse_genotypes</h3>
<blockquote class="blockquote">
<pre><code> parse_genotypes (genotypes_file)</code></pre>
</blockquote>
<div id="cell-23" class="cell">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> dgrec.example_data <span class="im">import</span> get_example_data_dir</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-24" class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>data_path<span class="op">=</span>get_example_data_dir()</span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>gen_list<span class="op">=</span>parse_genotypes(os.path.join(data_path,<span class="st">"sacB_genotypes.csv"</span>))</span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> g,n <span class="kw">in</span> itertools.islice(gen_list,<span class="dv">30</span>,<span class="dv">40</span>):</span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(n,<span class="st">"</span><span class="ch">\t</span><span class="st">"</span>,g)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>20   A72G,A79G
19   A72G,A79T,A91G
17   T67G,A91G
17   A76G,A79T
17   A68C,A72G
17   A111G
16   A68G,A91G
16   A86G,A91T
15   A72G,A91T
15   A79G,A86G</code></pre>
</div>
</div>
<hr>
<p><a href="https://github.com/dbikard/dgrec/blob/main/dgrec/utils.py#L165" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="get_aa_mut_list" class="level3">
<h3 class="anchored" data-anchor-id="get_aa_mut_list">get_aa_mut_list</h3>
<blockquote class="blockquote">
<pre><code> get_aa_mut_list (gen_list, refseq, frame=0, ori=1)</code></pre>
</blockquote>
<div id="cell-26" class="cell">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>read_ref_file<span class="op">=</span><span class="st">"sacB_ref.fasta"</span></span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a>refseq<span class="op">=</span><span class="bu">next</span>(SeqIO.parse(os.path.join(data_path,read_ref_file),<span class="st">"fasta"</span>))</span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a>refseq<span class="op">=</span><span class="bu">str</span>(refseq.seq)</span>
<span id="cb40-4"><a href="#cb40-4" aria-hidden="true" tabindex="-1"></a>aa_mut_list<span class="op">=</span>get_aa_mut_list(gen_list,refseq,ori<span class="op">=-</span><span class="dv">1</span>)</span>
<span id="cb40-5"><a href="#cb40-5" aria-hidden="true" tabindex="-1"></a>aa_mut_list[:<span class="dv">10</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>[('', 43341),
 ('Y22H', 351),
 ('H15Q', 277),
 ('D19E', 246),
 ('L17P', 200),
 ('V23A', 162),
 ('S11P', 117),
 ('D25E', 113),
 ('D19E,Y22H', 75),
 ('T16P', 61)]</code></pre>
</div>
</div>
<hr>
<p><a href="https://github.com/dbikard/dgrec/blob/main/dgrec/utils.py#L181" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="downsample_fastq_gz" class="level3">
<h3 class="anchored" data-anchor-id="downsample_fastq_gz">downsample_fastq_gz</h3>
<blockquote class="blockquote">
<pre><code> downsample_fastq_gz (input_file, output_file, num_reads=10000)</code></pre>
</blockquote>
<p>*Downsamples a compressed FASTQ file to the specified number of reads.</p>
<p>Args: input_file (str): Path to the input FASTQ.gz file. output_file (str): Path to the output FASTQ.gz file. num_reads (int, optional): Number of reads to keep. Defaults to 10000.*</p>
<div id="cell-28" class="cell">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a>input_file<span class="op">=</span>os.path.join(data_path,<span class="st">"sacB_example.fastq.gz"</span>)</span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a>output_file<span class="op">=</span><span class="st">"sacB_example_downsampled.fastq.gz"</span></span>
<span id="cb43-3"><a href="#cb43-3" aria-hidden="true" tabindex="-1"></a>downsample_fastq_gz(input_file, output_file, num_reads<span class="op">=</span><span class="dv">100</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<hr>
<p><a href="https://github.com/dbikard/dgrec/blob/main/dgrec/utils.py#L196" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="get_basename_without_extension" class="level3">
<h3 class="anchored" data-anchor-id="get_basename_without_extension">get_basename_without_extension</h3>
<blockquote class="blockquote">
<pre><code> get_basename_without_extension (file_path)</code></pre>
</blockquote>
<p>*Extracts the basename of a file without the extension.</p>
<p>Args: file_path (str): The path to the file.</p>
<p>Returns: str: The basename of the file without the extension.*</p>
<div id="cell-30" class="cell">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Example usage</span></span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a>file_path <span class="op">=</span> <span class="st">"C:/Users/John/Documents/my_file.txt"</span></span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a>basename_without_extension <span class="op">=</span> get_basename_without_extension(file_path)</span>
<span id="cb45-4"><a href="#cb45-4" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(basename_without_extension)  <span class="co"># Output: my_file</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>my_file</code></pre>
</div>
</div>
<hr>
<p><a href="https://github.com/dbikard/dgrec/blob/main/dgrec/utils.py#L216" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="pickle_save" class="level3">
<h3 class="anchored" data-anchor-id="pickle_save">pickle_save</h3>
<blockquote class="blockquote">
<pre><code> pickle_save (data_in, file_name_out)</code></pre>
</blockquote>
<hr>
<p><a href="https://github.com/dbikard/dgrec/blob/main/dgrec/utils.py#L224" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="pickle_load" class="level3">
<h3 class="anchored" data-anchor-id="pickle_load">pickle_load</h3>
<blockquote class="blockquote">
<pre><code> pickle_load (file_name_in)</code></pre>
</blockquote>
<hr>
<p><a href="https://github.com/dbikard/dgrec/blob/main/dgrec/utils.py#L230" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="make_dgr_oligos" class="level3">
<h3 class="anchored" data-anchor-id="make_dgr_oligos">make_dgr_oligos</h3>
<blockquote class="blockquote">
<pre><code> make_dgr_oligos (target:str, split_number:int)</code></pre>
</blockquote>
<p><em>Split the TR target into the input number and then generates the oligos to order</em></p>
<table class="caption-top table">
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>target</td>
<td>str</td>
<td>TR DNA</td>
</tr>
<tr class="even">
<td>split_number</td>
<td>int</td>
<td>Number of desired splits</td>
</tr>
</tbody>
</table>
<div id="cell-34" class="cell">
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a>target<span class="op">=</span><span class="st">'CCTCAGATACAAGCCGGCATAAATAATAACATATTCTATGACCATGATAATAGTGTAGGTGCAAACGCCAACGCTAAAAACACTGGAACCATGAACGGTAATACTGCAGGGACGAATATAGCCAAAACTTCT'</span></span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a>make_dgr_oligos(target,<span class="dv">4</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>['ATAACCTCAGATACAAGCCGGCATAAATAATAACA',
 'AATATGTTATTATTTATGCCGGCTTGTATCTGAGG',
 'TATTCTATGACCATGATAATAGTGTAGGTGCAA',
 'GCGTTTGCACCTACACTATTATCATGGTCATAG',
 'ACGCCAACGCTAAAAACACTGGAACCATGAACG',
 'TTACCGTTCATGGTTCCAGTGTTTTTAGCGTTG',
 'GTAATACTGCAGGGACGAATATAGCCAAAACTTCT',
 'CAGAAGAAGTTTTGGCTATATTCGTCCCTGCAGTA']</code></pre>
</div>
</div>
<hr>
<p><a href="https://github.com/dbikard/dgrec/blob/main/dgrec/utils.py#L278" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="reverse_comp_geno_list" class="level3">
<h3 class="anchored" data-anchor-id="reverse_comp_geno_list">reverse_comp_geno_list</h3>
<blockquote class="blockquote">
<pre><code> reverse_comp_geno_list (geno_list:list, ref_seq:str)</code></pre>
</blockquote>
<table class="caption-top table">
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>geno_list</td>
<td>list</td>
<td>List of genotypes</td>
</tr>
<tr class="even">
<td>ref_seq</td>
<td>str</td>
<td>string of the template sequence</td>
</tr>
</tbody>
</table>
<hr>
<p><a href="https://github.com/dbikard/dgrec/blob/main/dgrec/utils.py#L308" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="remove_position" class="level3">
<h3 class="anchored" data-anchor-id="remove_position">remove_position</h3>
<blockquote class="blockquote">
<pre><code> remove_position (geno, pos_list)</code></pre>
</blockquote>
<hr>
<p><a href="https://github.com/dbikard/dgrec/blob/main/dgrec/utils.py#L319" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="remove_position_list" class="level3">
<h3 class="anchored" data-anchor-id="remove_position_list">remove_position_list</h3>
<blockquote class="blockquote">
<pre><code> remove_position_list (geno_list, pos_list)</code></pre>
</blockquote>
<div id="cell-38" class="cell">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a>ref_genome<span class="op">=</span><span class="st">'AACGTATACGGCGGAATATTTGCCGAATGCCGTGTGGACGTAAGCGTGAACGTCAGGATCACGTTTCCCCGACCCGCTGGCATGTCAACAATACGGGAGAACACCTGTACCGCCTCGTTCGCCGCGC'</span></span>
<span id="cb55-2"><a href="#cb55-2" aria-hidden="true" tabindex="-1"></a>geno_list_test<span class="op">=</span>[(<span class="st">'T19C'</span>, <span class="dv">176012</span>),</span>
<span id="cb55-3"><a href="#cb55-3" aria-hidden="true" tabindex="-1"></a> (<span class="st">'T19C,T64A,T65A'</span>, <span class="dv">169</span>),</span>
<span id="cb55-4"><a href="#cb55-4" aria-hidden="true" tabindex="-1"></a> (<span class="st">'T19C,G36T'</span>, <span class="dv">40</span>),</span>
<span id="cb55-5"><a href="#cb55-5" aria-hidden="true" tabindex="-1"></a> (<span class="st">'T19C,T58G,T63A,T64G'</span>, <span class="dv">4</span>),</span>
<span id="cb55-6"><a href="#cb55-6" aria-hidden="true" tabindex="-1"></a> (<span class="st">'T19C,A42C'</span>, <span class="dv">14</span>),</span>
<span id="cb55-7"><a href="#cb55-7" aria-hidden="true" tabindex="-1"></a> (<span class="st">'T19C,T63G'</span>, <span class="dv">13</span>),</span>
<span id="cb55-8"><a href="#cb55-8" aria-hidden="true" tabindex="-1"></a> (<span class="st">'T19C,T52A,T64A,T65A'</span>, <span class="dv">19</span>),</span>
<span id="cb55-9"><a href="#cb55-9" aria-hidden="true" tabindex="-1"></a> (<span class="st">'T19C,A41C,A57C,T58C'</span>, <span class="dv">1</span>),</span>
<span id="cb55-10"><a href="#cb55-10" aria-hidden="true" tabindex="-1"></a> (<span class="st">'T19C,T52A'</span>, <span class="dv">94</span>),</span>
<span id="cb55-11"><a href="#cb55-11" aria-hidden="true" tabindex="-1"></a> (<span class="st">'T19C,T64A,T65G'</span>, <span class="dv">214</span>),</span>
<span id="cb55-12"><a href="#cb55-12" aria-hidden="true" tabindex="-1"></a> (<span class="st">'T19C,T63A,T64C,T65A'</span>, <span class="dv">2</span>),</span>
<span id="cb55-13"><a href="#cb55-13" aria-hidden="true" tabindex="-1"></a> (<span class="st">'T19C,A49C,T64C,T91C'</span>, <span class="dv">2</span>),</span>
<span id="cb55-14"><a href="#cb55-14" aria-hidden="true" tabindex="-1"></a> (<span class="st">'T19C,T32C,T52C,T64A,T65G,T84C,T91A'</span>, <span class="dv">1</span>),</span>
<span id="cb55-15"><a href="#cb55-15" aria-hidden="true" tabindex="-1"></a> (<span class="st">'T19C,T82C,T84A'</span>, <span class="dv">8</span>),</span>
<span id="cb55-16"><a href="#cb55-16" aria-hidden="true" tabindex="-1"></a> (<span class="st">'T19C,T64C'</span>, <span class="dv">308</span>),</span>
<span id="cb55-17"><a href="#cb55-17" aria-hidden="true" tabindex="-1"></a> (<span class="st">'T19C,T40C,G43A,T58A,A71C,T91C'</span>, <span class="dv">1</span>),</span>
<span id="cb55-18"><a href="#cb55-18" aria-hidden="true" tabindex="-1"></a> (<span class="st">'T52A,T77C'</span>, <span class="dv">1</span>),</span>
<span id="cb55-19"><a href="#cb55-19" aria-hidden="true" tabindex="-1"></a> (<span class="st">'T19C,T52C,T58A'</span>, <span class="dv">9</span>),</span>
<span id="cb55-20"><a href="#cb55-20" aria-hidden="true" tabindex="-1"></a> (<span class="st">'T20C,-52C,T58C,A71C,T77C'</span>, <span class="dv">1</span>),</span>
<span id="cb55-21"><a href="#cb55-21" aria-hidden="true" tabindex="-1"></a> (<span class="st">'T19C,G70T'</span>, <span class="dv">110</span>),</span>
<span id="cb55-22"><a href="#cb55-22" aria-hidden="true" tabindex="-1"></a> (<span class="st">'T19C,T32C,T65C,T82C'</span>, <span class="dv">1</span>),</span>
<span id="cb55-23"><a href="#cb55-23" aria-hidden="true" tabindex="-1"></a> (<span class="st">'T19C,T32C,T46C'</span>, <span class="dv">4</span>),</span>
<span id="cb55-24"><a href="#cb55-24" aria-hidden="true" tabindex="-1"></a> (<span class="st">'T19C,T32C,A41C,A49C,T64A,T65G'</span>, <span class="dv">1</span>),</span>
<span id="cb55-25"><a href="#cb55-25" aria-hidden="true" tabindex="-1"></a> (<span class="st">'T19C,T64A'</span>, <span class="dv">115</span>),</span>
<span id="cb55-26"><a href="#cb55-26" aria-hidden="true" tabindex="-1"></a> (<span class="st">'T19C,T58C'</span>, <span class="dv">68</span>)]</span>
<span id="cb55-27"><a href="#cb55-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-28"><a href="#cb55-28" aria-hidden="true" tabindex="-1"></a>geno_list_test<span class="op">=</span>remove_position_list(geno_list_test,[<span class="dv">19</span>])</span>
<span id="cb55-29"><a href="#cb55-29" aria-hidden="true" tabindex="-1"></a>reverse_comp_geno_list(geno_list_test,ref_genome)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>[('', 176012),
 ('A61T,A62T', 169),
 ('C90A', 40),
 ('A62C,A63T,A68C', 4),
 ('T84G', 14),
 ('A63C', 13),
 ('A61T,A62T,A74T', 19),
 ('A68G,T69G,T85G', 1),
 ('A74T', 94),
 ('A61C,A62T', 214),
 ('A61T,A62G,A63T', 2),
 ('A35G,A62G,T77G', 2),
 ('A35T,A42G,A61C,A62T,A74G,A94G', 1),
 ('A42T,A44G', 8),
 ('A62G', 308),
 ('A35G,T55G,A68T,C83T,A86G', 1),
 ('A49G,A74T', 1),
 ('A68T,A74G', 9),
 ('A49G,T55G,A68G,-74G,A106G', 1),
 ('C56A', 110),
 ('A44G,A61G,A94G', 1),
 ('A80G,A94G', 4),
 ('A61C,A62T,T77G,T85G,A94G', 1),
 ('A62T', 115),
 ('A68G', 68)]</code></pre>
</div>
</div>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
        const codeEl = trigger.previousElementSibling.cloneNode(true);
        for (const childEl of codeEl.children) {
          if (isCodeAnnotation(childEl)) {
            childEl.remove();
          }
        }
        return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp("https:\/\/dbikard\.github\.io\/dgrec");
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->




<footer class="footer"><div class="nav-footer"><div class="nav-footer-center"><div class="toc-actions d-sm-block d-md-none"><ul><li><a href="https://github.com/dbikard/dgrec/issues/new" class="toc-action"><i class="bi bi-github"></i>Report an issue</a></li></ul></div></div></div></footer></body></html>